<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data structures, clustering and testing for single 
    cell immune receptor repertoires (scRNAseq RepSeq/AIRR-seq) • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Data structures, clustering and testing for single 
    cell immune receptor repertoires (scRNAseq RepSeq/AIRR-seq)">
<meta property="og:description" content="Methods to cluster and analyze high-throughput
    single cell immune cell repertoires, especially from the 10X Genomics
    VDJ solution. Contains an R interface to CD-HIT (Li and Godzik 2006).   
    Methods to visualize and analyze paired heavy-light chain data.
    Tests for specific expansion, as well as omnibus oligoclonality under
    hypergeometric models.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/cdr3_clustering.html">Clustering and differential usage of repertoire CDR3 sequences</a>
    </li>
    <li>
      <a href="articles/cr-overview.html">An Introduction to CellaRepertorium</a>
    </li>
    <li>
      <a href="articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="cellarepertorium" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#cellarepertorium" class="anchor"></a>CellaRepertorium</h1></div>
<p>This package contains methods for clustering, pairing and testing single cell RepSeq data, especially as generated by <a href="https://support.10xgenomics.com/single-cell-vdj">10X genomics VDJ solution</a>.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Upon acceptance in Bioconductor, install with</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">'BiocManager'</span>) <span class="co"># if you don't have it yet</span>
BiocManager::<span class="kw">install</span>(<span class="st">'CellaRepertorium'</span>)</code></pre></div>
<p>For the development version, install via <code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github('amcdavid/CellaRepertorium')</a></code>.</p>
</div>
<div id="data-requirements-and-package-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#data-requirements-and-package-structure" class="anchor"></a>Data requirements and package structure</h2>
<p>The fundamental unit this package operates on is the <strong>contig</strong>, which is a section of contiguously stitched reads from a <strong>single cell</strong>. Each contig belongs to one (and only one) cell, however, cells generate multiple contigs.<br><img src="reference/figures/contig_schematic.png"></p>
<p>Contigs can also belong to a <strong>cluster</strong>. Because of these two many-to-one mappings, these data can be thought as a series of ragged arrays. The links between them mean they are relational data. A <code><a href="reference/ContigCellDB.html">ContigCellDB()</a></code> object wraps each of these objects as a sequence of three <code>data.frames</code> (<code><a href="https://tibble.tidyverse.org/reference/tibble.html">dplyr::tibble()</a></code>, actually). <code><a href="reference/ContigCellDB.html">ContigCellDB()</a></code> also tracks columns (the primary keys) that uniquely identify each row in each of these tables. The <code>contig_tbl</code> is the <code>tibble</code> containing <strong>contigs</strong>, the <code>cell_tbl</code> contains the <strong>cells</strong>, and the <code>cluster_tbl</code> contains the <strong>clusters</strong>.</p>
<p>The <code>contig_pk</code>, <code>cell_pk</code> and <code>cluster_pk</code> specify the columns that identify a contig, cell and cluster, respectively. These will serve as foreign keys that link the three tables together. The tables are kept in sync so that subsetting the contigs will subset the cells, and clusters, and vice-versa.</p>
<p><img src="reference/figures/table_schematic.png"></p>
<p>Of course, each of these tables can contain many other columns that will serve as covariates for various analyses, such as the <strong>CDR3</strong> sequence, or the identity of the <strong>V</strong>, <strong>D</strong> and <strong>J</strong> regions. Various derived quantities that describe cells and clusters can also be calculated, and added to these tables, such as the <strong>medoid</strong> of a cluster – a contig that minimizes the average distance to all other clusters.</p>
</div>
<div id="some-functions-of-interest" class="section level2">
<h2 class="hasAnchor">
<a href="#some-functions-of-interest" class="anchor"></a>Some functions of interest</h2>
<p>Mainly, this package seeks to enforce proper schema of single cell repertoire data and stay out the user’s way for various summaries they might conduct.</p>
<p>However, there are a variety of specialized functions, as well:</p>
<ul>
<li>
<code><a href="reference/cdhit_ccdb.html">cdhit_ccdb()</a></code>: An R interface to CDhit, which was originally ported by Thomas Lin Pedersen.</li>
<li>
<code><a href="reference/fine_clustering.html">fine_clustering()</a></code>: clustering CDR3 by edit distances (possibly using empirical amino acid substitution matrices)</li>
<li>
<code><a href="reference/canonicalize_cell.html">canonicalize_cell()</a></code>: Return a single contig for each cell, e.g., for combining VDJ information with 5’-based single cell expression</li>
<li>
<code><a href="reference/ccdb_join.html">ccdb_join()</a></code>: join a <code>ccdb</code> object from this package to a <code>SingleCellExperiment</code> object, by droplet barcode.</li>
<li>
<code><a href="reference/cluster_permute_test.html">cluster_permute_test()</a></code>: permutation tests of cluster statistics</li>
<li>
<code><a href="reference/cluster_test_by.html">cluster_logistic_test()</a></code>: logistic regression tests for overrepresentation of clusters among cells</li>
<li>
<code><a href="reference/pairing_tables.html">pairing_tables()</a></code>: Generate pairings of contigs within each cell in a way that they can be plotted</li>
</ul>
</div>
<div id="interfacing-related-packages-for-clonal-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#interfacing-related-packages-for-clonal-analyses" class="anchor"></a>Interfacing related packages for clonal analyses</h2>
<ul>
<li>To combine repertoire information with expression of endogenous mRNAs, this package has been used with <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment::SingleCellExperiment()</a></code> and <a href="https://satijalab.org/seurat/">Seurat</a> after generating cell canonicalizations.</li>
<li>Many tools from the <a href="https://alakazam.readthedocs.io/en/version-0.2.11/">Immcantation</a> suite can work directly on <code><a href="reference/ContigCellDB.html">ContigCellDB()</a></code> objects.</li>
</ul>
</div>
</div>
<div id="acknowledgments" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgments" class="anchor"></a>Acknowledgments</h1>
<p>Development of CellaRepertorium was funded in part by UL1 TR002001 (PI Bennet/Zand) pilot to Andrew McDavid.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/CellaRepertorium">https://​www.bioconductor.org/​packages/​CellaRepertorium</a>
</li>
<li>Browse source code at <br><a href="https://github.com/amcdavid/CellaRepertorium/">https://​github.com/​amcdavid/​CellaRepertorium/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/amcdavid/CellaRepertorium/issues">https://​github.com/​amcdavid/​CellaRepertorium/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Andrew McDavid <br><small class="roles"> Author, maintainer </small>  </li>
<li>Yu Gu <br><small class="roles"> Author </small>  </li>
<li>Erik VonKaenel <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Yu Gu, Erik VonKaenel.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
