<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering and differential usage of repertoire CDR3 sequences • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering and differential usage of repertoire CDR3 sequences">
<meta property="og:description" content="CellaRepertorium">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering and differential usage of repertoire CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/cr-overview.html">An Introduction to CellaRepertorium</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Clustering and differential usage of repertoire
CDR3 sequences</h1>
                        <h4 data-toc-skip class="author">Andrew
McDavid</h4>
            <address class="author_afil">
      University of Rochester, Department of Biostatistics and
Computational
Biology<br><a class="author_email" href="mailto:#"></a><a href="mailto:Andrew_McDavid@urmc.rochester.edu" class="email">Andrew_McDavid@urmc.rochester.edu</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/HEAD/vignettes/cdr3_clustering.Rmd" class="external-link"><code>vignettes/cdr3_clustering.Rmd</code></a></small>
      <div class="hidden name"><code>cdr3_clustering.Rmd</code></div>

    </div>

    
    
<p>In this vignette we demonstrate clustering of 3rd complementary
determining region sequence (CDR3) and V-J gene identity of mouse T
cells, ways to visualize and explore clusters that are expanded, pairing
of alpha-beta clusters, tests of differential CDR3 usage, and
permutation tests for overall clonal properties.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/amcdavid/CellaRepertorium" class="external-link">CellaRepertorium</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org" class="external-link">purrr</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="load-filtered-contig-files">Load filtered contig files<a class="anchor" aria-label="anchor" href="#load-filtered-contig-files"></a>
</h2>
<p>We begin with a <code>data.frame</code> of concatenated contig files
(‘all_contig_annotations.csv’), output from the Cellranger VDJ
pipeline.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">contigs_qc</span><span class="op">)</span>
<span class="va">MIN_CDR3_AA</span> <span class="op">=</span> <span class="fl">6</span>


<span class="va">cdb</span> <span class="op">=</span> <span class="fu"><a href="../reference/ContigCellDB-fun.html">ContigCellDB_10XVDJ</a></span><span class="op">(</span><span class="va">contigs_qc</span>, contig_pk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>, <span class="st">'contig_id'</span><span class="op">)</span>, cell_pk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span><span class="op">)</span><span class="op">)</span>
<span class="va">cdb</span>
<span class="co">#&gt; ContigCellDB of 1508 contigs; 832 cells; and 0 clusters.</span>
<span class="co">#&gt; Contigs keyed by barcode, pop, sample, contig_id; cells keyed by barcode, pop, sample.</span></code></pre></div>
<p>Initially we start with 832 cells and 1508 contigs. We keep contigs
that are</p>
<ul>
<li>full - length</li>
<li>productive</li>
<li>high-confidence</li>
<li>only from T cells</li>
<li>and with CDR3 sufficiently long.</li>
</ul>
<p>Then we add a descriptive readable name for each contig.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdb</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">$</span><span class="va">contig_tbl</span>, <span class="va">full_length</span>, <span class="va">productive</span> <span class="op">==</span> <span class="st">'True'</span>, <span class="va">high_confidence</span>, <span class="va">chain</span> <span class="op">!=</span> <span class="st">'Multi'</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_length.html" class="external-link">str_length</a></span><span class="op">(</span><span class="va">cdr3</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">MIN_CDR3_AA</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span> fancy_name <span class="op">=</span> <span class="fu"><a href="../reference/fancy_name_contigs.html">fancy_name_contigs</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">pop</span>, <span class="st">'_'</span>, <span class="va">sample</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>After filtering, there are 832 cells and 1496 contigs.</p>
</div>
<div class="section level2">
<h2 id="clustering-contigs-by-sequence-characteristics">Clustering contigs by sequence characteristics<a class="anchor" aria-label="anchor" href="#clustering-contigs-by-sequence-characteristics"></a>
</h2>
<p>As a first step to define clonotypes, we will first find equivalence
classes of CDR3 sequences with the program <a href="http://weizhongli-lab.org/cdhit_suite/cgi-bin/index.cgi?cmd=cd-hit" class="external-link">CD-HIT</a>.
In this case, we use the translated amino acid residues, but often one
might prefer to use the DNA sequences, by setting the
<code>sequence_key</code> accordingly and <code>type = 'DNA'</code>.
Additionally, a higher identity threshold might be appropriate (see
below).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span><span class="op">(</span><span class="va">cdb</span>, sequence_key <span class="op">=</span> <span class="st">'cdr3'</span>, type <span class="op">=</span> <span class="st">'AA'</span>, cluster_pk <span class="op">=</span> <span class="st">'aa80'</span>, 
                  identity <span class="op">=</span> <span class="fl">.8</span>, min_length <span class="op">=</span> <span class="fl">5</span>, G <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span><span class="op">(</span><span class="va">aa80</span>, sequence_key <span class="op">=</span> <span class="st">'cdr3'</span>, type <span class="op">=</span> <span class="st">'AA'</span>, keep_clustering_details <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Calculating intradistances on 988 clusters.</span>
<span class="co">#&gt; Summarizing</span></code></pre></div>
<p>This partitions sequences into sets with &gt;80% mutual similarity in
the amino acid sequence, adds some additional information about the
clustering, and returns it as a <code>ContigCellDB</code> object named
<code>aa80</code>. The primary key for the clusters is aa80. The
<code>min_length</code> can be set somewhat smaller, but there is a
lower limit for the cdhit algorithm. <code>G=1</code>, the default,
specifies a global alignment. This is almost always what is desired, but
local alignment is available if <code>G=0</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">cluster_tbl</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">#&gt;    aa80 avg_distance fc               n_cluster</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     3            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         2</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     4            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     5            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     6            0 <span style="color: #949494;">&lt;named list [5]&gt;</span>         1</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">contig_id</span>, <span class="va">aa80</span>, <span class="va">is_medoid</span>, <span class="va">`d(medoid)`</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">#&gt;   contig_id                    aa80 is_medoid `d(medoid)`</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ATCTACTCAGTATGCT-1_contig_3     1 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ACTGTCCTCAATCACG-1_contig_3     2 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> CACCTTGTCCAATGGT-1_contig_2     3 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CACCTTGTCCAATGGT-1_contig_2     3 FALSE               0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> CGGACGTGTTCATGGT-1_contig_1     4 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> CTGCTGTTCCCTAATT-1_contig_4     5 TRUE                0</span></code></pre></div>
<p>The <code>cluster_tbl</code> lists the 988 80% identity groups found,
including the number of contigs in the cluster, and the average distance
between elements in the group. In the <code>contig_tbl</code>, there are
two columns specifying if the contig <code>is_medoid</code>, that is, is
the most representative element of the set and the distance to the
medoid element <code>d(medoid)</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cluster_plot.html">cluster_plot</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: cowplot</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="section level3">
<h3 id="cluster-cdr3-dna-sequences">Cluster CDR3 DNA sequences<a class="anchor" aria-label="anchor" href="#cluster-cdr3-dna-sequences"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdb</span> <span class="op">=</span> <span class="fu"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span><span class="op">(</span><span class="va">cdb</span>, <span class="st">'cdr3_nt'</span>, type <span class="op">=</span> <span class="st">'DNA'</span>, cluster_pk <span class="op">=</span> <span class="st">'DNA97'</span>, identity <span class="op">=</span> <span class="fl">.965</span>, min_length <span class="op">=</span> <span class="va">MIN_CDR3_AA</span><span class="op">*</span><span class="fl">3</span><span class="op">-</span><span class="fl">1</span>, G <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">cdb</span> <span class="op">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span><span class="op">(</span><span class="va">cdb</span>, sequence_key <span class="op">=</span> <span class="st">'cdr3_nt'</span>, type <span class="op">=</span> <span class="st">'DNA'</span><span class="op">)</span>
<span class="co">#&gt; Calculating intradistances on 1342 clusters.</span>
<span class="co">#&gt; Summarizing</span>

<span class="fu"><a href="../reference/cluster_plot.html">cluster_plot</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">)</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We can also cluster by DNA identity.</p>
</div>
<div class="section level3">
<h3 id="cluster-by-v-j-identity">Cluster by V-J identity<a class="anchor" aria-label="anchor" href="#cluster-by-v-j-identity"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">germline_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_germline.html">cluster_germline</a></span><span class="op">(</span><span class="va">cdb</span>, segment_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'chain'</span><span class="op">)</span>, cluster_pk <span class="op">=</span> <span class="st">'segment_idx'</span><span class="op">)</span>
<span class="co">#&gt; Warning in replace_cluster_tbl(ccdb, cluster_tbl, cl_con_tbl, cluster_pk =</span>
<span class="co">#&gt; cluster_pk): Replacing `cluster_tbl` with DNA97.</span></code></pre></div>
<p>We can cluster by any other feature of the contigs. Here we cluster
each contig based on the chain and V-J genes. This gives us the set of
observed V-J pairings:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">germline_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span><span class="op">(</span><span class="va">germline_cluster</span>, sequence_key <span class="op">=</span> <span class="st">'cdr3_nt'</span>, type <span class="op">=</span> <span class="st">'DNA'</span><span class="op">)</span>
<span class="co">#&gt; Calculating intradistances on 700 clusters.</span>
<span class="co">#&gt; Summarizing</span>
<span class="co">#&gt; Warning in left_join_warn(d_medoid, contig_tbl, by = ccdb$contig_pk, overwrite =</span>
<span class="co">#&gt; TRUE): Overwriting fields d(medoid), is_medoid in table contig_tbl</span>
<span class="fu"><a href="../reference/mutate_cdb.html">filter_cdb</a></span><span class="op">(</span><span class="va">germline_cluster</span>, <span class="va">chain</span> <span class="op">==</span> <span class="st">'TRB'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/plot_cluster_factors.html">plot_cluster_factors</a></span><span class="op">(</span>factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'v_gene'</span>,<span class="st">'j_gene'</span><span class="op">)</span>, statistic <span class="op">=</span> <span class="st">'contigs'</span>, type <span class="op">=</span> <span class="st">'heatmap'</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Number of pairs. The pearson residual (showing the difference from
expected counts given marginals) is probably more informative, set
<code>statistic = 'residual'</code> for this.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">germline_cluster</span><span class="op">$</span><span class="va">cluster_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chain</span> <span class="op">==</span> <span class="st">'TRB'</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">v_gene</span>, y <span class="op">=</span> <span class="va">j_gene</span>, fill <span class="op">=</span> <span class="va">avg_distance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Average Levenshtein distance of CDR3 within each pair. This might be
turned into a z-score by fitting a weighted linear model with
sum-to-zero contrasts and returning the studentized residuals. This
could determine if a pairing has an unexpected small, or large, within
cluster distance.</p>
</div>
<div class="section level3">
<h3 id="expanded-clusters">Expanded clusters<a class="anchor" aria-label="anchor" href="#expanded-clusters"></a>
</h3>
<p>Next, we will examine the clusters that are found in many contigs.
First we will get a canonical contig to represent each cluster. This
will be the medoid contig, by default.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/canonicalize_cluster.html">canonicalize_cluster</a></span><span class="op">(</span><span class="va">aa80</span>, representative <span class="op">=</span> <span class="st">'cdr3'</span>, contig_fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cdr3'</span>, <span class="st">'cdr3_nt'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'d_gene'</span>, <span class="st">'j_gene'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Filtering `contig_tbl` by `is_medoid`, override by setting `contig_filter_args == TRUE`</span></code></pre></div>
<p><code>aa80</code> now includes the fields listed in
<code>contig_fields</code> in the <code>cluster_tbl</code>, using the
values found in the medoid contig.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MIN_OLIGO</span> <span class="op">=</span> <span class="fl">7</span>
<span class="va">oligo_clusters</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">cluster_tbl</span>, <span class="va">n_cluster</span> <span class="op">&gt;=</span> <span class="va">MIN_OLIGO</span><span class="op">)</span>
<span class="va">oligo_contigs</span> <span class="op">=</span> <span class="va">aa80</span>
<span class="va">oligo_contigs</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">semi_join</a></span><span class="op">(</span><span class="va">oligo_contigs</span><span class="op">$</span><span class="va">contig_tbl</span>, <span class="va">oligo_clusters</span>, by <span class="op">=</span> <span class="st">'aa80'</span><span class="op">)</span>
<span class="va">oligo_contigs</span>
<span class="co">#&gt; ContigCellDB of 54 contigs; 832 cells; and 4 clusters.</span>
<span class="co">#&gt; Contigs keyed by barcode, pop, sample, contig_id; cells keyed by barcode, pop, sample.</span></code></pre></div>
<p>Get contigs/cells/clusters found at least 7 times (across contigs).
Note that replacing <code>contig_tbl</code> with the subset selected
with the <code>semi_join</code> also automatically subsetted the
<code>cell_tbl</code> and <code>cluster_tbl</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">oligo_clusters</span> <span class="op">=</span> <span class="va">oligo_contigs</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>`n subjects observed` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">oligo_clusters</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "aa80"</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">oligo_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">:</span><span class="va">cdr3</span>, <span class="va">chain</span><span class="op">:</span><span class="va">j_gene</span>, <span class="va">avg_distance</span>, <span class="va">n_cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="5%">
<col width="21%">
<col width="17%">
<col width="6%">
<col width="10%">
<col width="7%">
<col width="7%">
<col width="13%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="right">aa80</th>
<th align="right">n subjects observed</th>
<th align="left">cdr3</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="right">avg_distance</th>
<th align="right">n_cluster</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">111</td>
<td align="right">6</td>
<td align="left">CVVGDRGSALGRLHF</td>
<td align="left">TRA</td>
<td align="left">TRAV11</td>
<td align="left">None</td>
<td align="left">TRAJ18</td>
<td align="right">0.6071429</td>
<td align="right">28</td>
</tr>
<tr class="even">
<td align="right">172</td>
<td align="right">5</td>
<td align="left">CAVSRASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV9N-3</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">2.1111111</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="right">296</td>
<td align="right">6</td>
<td align="left">CAASASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV14D-2</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">1.5000000</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="right">808</td>
<td align="right">4</td>
<td align="left">CATGNYAQGLTF</td>
<td align="left">TRA</td>
<td align="left">TRAV8D-2</td>
<td align="left">None</td>
<td align="left">TRAJ26</td>
<td align="right">1.3333333</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p>Report some statistics about these expanded clusters, such as how
often they are found, how many subjects, etc.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">oligo_plot</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">oligo_contigs</span><span class="op">$</span><span class="va">contig_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">representative</span>, fill <span class="op">=</span> <span class="va">chain</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">'qual'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">oligo_plot</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>These always come from a single chain.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">oligo_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span>   <span class="va">sample</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">pop</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>But come from multiple populations and samples.</p>
</div>
<div class="section level3">
<h3 id="some-simple-phylogenetic-relationships">Some simple phylogenetic relationships<a class="anchor" aria-label="anchor" href="#some-simple-phylogenetic-relationships"></a>
</h3>
<p>By using the within-cluster distances, some rudamentory plots
attempting to show phylogenetic associations are possible. (These are
most biologically appropriate for B cells that undergo somatic
hypermutation.)</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro" class="external-link">ggdendro</a></span><span class="op">)</span>

<span class="va">dendro_plot</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ccdb</span>, <span class="va">idx</span>, <span class="va">method</span> <span class="op">=</span> <span class="st">'complete'</span><span class="op">)</span><span class="op">{</span>
    <span class="va">h</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ccdb</span><span class="op">$</span><span class="va">cluster_tbl</span>, <span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/tidyeval-compat.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">ccdb</span><span class="op">$</span><span class="va">cluster_pk</span><span class="op">)</span> <span class="op">==</span> <span class="va">idx</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">fc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
    <span class="va">quer</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ccdb</span><span class="op">$</span><span class="va">contig_tbl</span>, <span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/tidyeval-compat.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">ccdb</span><span class="op">$</span><span class="va">cluster_pk</span><span class="op">)</span> <span class="op">==</span> <span class="va">idx</span><span class="op">)</span>
    <span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="va">h</span><span class="op">$</span><span class="va">distance_mat</span><span class="op">)</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/dendro_data.html" class="external-link">dendro_data</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"rectangle"</span><span class="op">)</span>
    <span class="va">hc</span><span class="op">$</span><span class="va">labels</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">hc</span><span class="op">$</span><span class="va">labels</span>, <span class="va">quer</span><span class="op">)</span>
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">hc</span><span class="op">$</span><span class="va">segments</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xend<span class="op">=</span><span class="va">xend</span>, yend<span class="op">=</span><span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">hc</span><span class="op">$</span><span class="va">labels</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">sample</span>, label <span class="op">=</span> <span class="va">fancy_name</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span>, angle <span class="op">=</span> <span class="fl">60</span>, hjust <span class="op">=</span><span class="fl">0</span>, vjust <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'AA Distance'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">to_plot</span> <span class="op">=</span> <span class="va">aa80</span><span class="op">$</span><span class="va">cluster_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/ranking.html" class="external-link">min_rank</a></span><span class="op">(</span><span class="op">-</span><span class="va">n_cluster</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>

<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">$</span><span class="va">aa80</span>, <span class="op">~</span> <span class="fu">dendro_plot</span><span class="op">(</span><span class="va">aa80</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>A full-blown generative model of clonal generation and selection
would be recommended for any actual analysis, but these plots may
suffice to get a quick idea of the phylogenetic structure.</p>
</div>
<div class="section level3">
<h3 id="formal-testing-for-frequency-differences">Formal testing for frequency differences<a class="anchor" aria-label="anchor" href="#formal-testing-for-frequency-differences"></a>
</h3>
<p>We can test for differential usage of a clone, or cluster with
<code>cluster_logistic_test</code> and <code>cluster_test_by</code>. The
latter splits the <code>cluster_tbl</code> by
<code>field = 'chain'</code>, thereby adjusting the number of cell
trials included in the “denominator” of the logistic regression. The
formula tests for differences between populations, including the sample
as a random effect, and only tests clusters that are included in the
<code>oligo_clusters</code> set.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_out</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_logistic_test.html">cluster_test_by</a></span><span class="op">(</span><span class="va">aa80</span>, fields <span class="op">=</span> <span class="st">'chain'</span>, tbl <span class="op">=</span> <span class="st">'cluster_tbl'</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">pop</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">sample</span><span class="op">)</span>, filterset <span class="op">=</span> <span class="fu"><a href="../reference/cluster_filterset.html">cluster_filterset</a></span><span class="op">(</span>white_list <span class="op">=</span> <span class="va">oligo_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">oligo_clusters</span><span class="op">)</span>
<span class="co">#&gt; Fitting mixed logistic models to 4 clusters.</span>
<span class="co">#&gt; Loading required namespace: broom</span>
<span class="co">#&gt; Loading required namespace: lme4</span>
<span class="co">#&gt; Loading required namespace: broom.mixed</span>
<span class="co">#&gt; boundary (singular) fit: see help('isSingular')</span>
<span class="co">#&gt; boundary (singular) fit: see help('isSingular')</span>
<span class="co">#&gt; boundary (singular) fit: see help('isSingular')</span>
<span class="co">#&gt; Fitting mixed logistic models to 0 clusters.</span>
<span class="co">#&gt; Joining, by = c("chain", "aa80")</span>

<span class="va">mm_out</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">mm_out</span>, conf.low <span class="op">=</span> <span class="va">estimate</span><span class="op">-</span><span class="fl">1.96</span><span class="op">*</span><span class="va">std.error</span>, 
                conf.high <span class="op">=</span> <span class="va">estimate</span> <span class="op">+</span> <span class="fl">1.96</span><span class="op">*</span><span class="va">std.error</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm_outj</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="va">mm_out</span><span class="op">)</span>, <span class="va">term</span> <span class="op">==</span> <span class="st">'popbalbc'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">representative</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mm_outj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">representative</span>, ymin <span class="op">=</span> <span class="va">conf.low</span>, ymax <span class="op">=</span> <span class="va">conf.high</span>, y <span class="op">=</span> <span class="va">estimate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Isomorph"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"log odds of isomorph"</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/per_iso_tests-1.png" width="700"></p>
<p>We test if the binomial rate of clone expression differs between
balbc and b6, for the selected clones. None appear to be different.</p>
</div>
<div class="section level3">
<h3 id="length-of-cdr3">Length of CDR3<a class="anchor" aria-label="anchor" href="#length-of-cdr3"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op">=</span> <span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>cdr3_length <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_length.html" class="external-link">str_length</a></span><span class="op">(</span><span class="va">cdr3_nt</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">pop</span>, x<span class="op">=</span> <span class="va">cdr3_length</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">'qual'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">sample</span> <span class="op">~</span><span class="va">chain</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">55</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>Some authors have noted that the length of the CDR3 region can be
predictive of T cell differentiation. In our study, there doesn’t appear
to be a noticeable difference between BALB/c and C57BL/6J (b6) mice, but
if we needed to make sure, an appropriate procedure would be to run a
mixed model with a random <code>sample</code> effect (assumed to
represent a biological replicate).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdr_len</span> <span class="op">=</span> <span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">chain</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html" class="external-link">do</a></span><span class="op">(</span><span class="fu">broom</span><span class="fu">::</span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="fu">lme4</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html" class="external-link">lmer</a></span><span class="op">(</span><span class="va">cdr3_length</span> <span class="op">~</span> <span class="va">pop</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">sample</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">.</span><span class="op">)</span>, conf.int <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; boundary (singular) fit: see help('isSingular')</span>
<span class="co">#&gt; boundary (singular) fit: see help('isSingular')</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cdr_len</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">'popbalbc'</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html" class="external-link">interaction</a></span><span class="op">(</span><span class="va">chain</span>, <span class="va">term</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">estimate</span>, ymin <span class="op">=</span> <span class="va">conf.low</span>, ymax <span class="op">=</span> <span class="va">conf.high</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Length(CDR3 Nt)'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Term/Chain'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/cdr3_len-1.png" width="288"></p>
<p>We end up with a (harmless) convergence warning about a singular fit.
This is expected, because the <code>samples</code> aren’t actually
replicates – they are just subsamples drawn for illustrative purposes.
The Balbc mice have .5 fewer nucleotides per contig, on average, and
this is not significant.</p>
</div>
</div>
<div class="section level2">
<h2 id="clonal-pairs">Clonal pairs<a class="anchor" aria-label="anchor" href="#clonal-pairs"></a>
</h2>
<p>Next, we can examine the pairing between <span class="math inline">\(\alpha-\beta\)</span> chains and see if any pairs
are found more than once.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span><span class="op">$</span><span class="va">cluster_pk</span> <span class="op">=</span> <span class="st">'representative'</span>
<span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/rank_prevalence_ccdb.html">rank_prevalence_ccdb</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">)</span>
<span class="va">pairing_list</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairing_tables.html">pairing_tables</a></span><span class="op">(</span><span class="va">aa80</span>, table_order <span class="op">=</span> <span class="fl">2</span>, orphan_level <span class="op">=</span> <span class="fl">1</span>, min_expansion <span class="op">=</span> <span class="fl">3</span>, cluster_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><code>pairing_tables</code> finds all contig combinations of order
<code>table_order</code> across cells. Among those combinations that
occur at least <code>min_expansion</code> times, the expanded
combinations and and any other combinations that shared an expanded
combo.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pairing_list</span><span class="op">$</span><span class="va">cell_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cluster_idx.1_fct</span>, y <span class="op">=</span> <span class="va">cluster_idx.2_fct</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">sample</span>, shape <span class="op">=</span> <span class="va">pop</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.2</span>, height <span class="op">=</span> <span class="fl">.2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'TRB'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'TRA'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span><span class="op">)</span><span class="op">)</span>

<span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_axis_labels.html">map_axis_labels</a></span><span class="op">(</span><span class="va">pairs_plt</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx1_tbl</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx2_tbl</span>, aes_label  <span class="op">=</span> <span class="st">'chain'</span><span class="op">)</span>
<span class="va">pairs_plt</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/plot_expanded-1.png" width="700"></p>
<div class="section level3">
<h3 id="expanded-clones">Expanded clones<a class="anchor" aria-label="anchor" href="#expanded-clones"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">whitelist</span> <span class="op">=</span> <span class="va">oligo_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>cluster_idx.1 <span class="op">=</span> <span class="va">representative</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">pairing_list</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairing_tables.html">pairing_tables</a></span><span class="op">(</span><span class="va">aa80</span>, table_order <span class="op">=</span> <span class="fl">2</span>, orphan_level <span class="op">=</span> <span class="fl">1</span>, min_expansion <span class="op">=</span> <span class="cn">Inf</span>, cluster_whitelist <span class="op">=</span> <span class="va">whitelist</span>,  cluster_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span><span class="op">)</span><span class="op">)</span>

<span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pairing_list</span><span class="op">$</span><span class="va">cell_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cluster_idx.1_fct</span>, y <span class="op">=</span> <span class="va">cluster_idx.2_fct</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">sample</span>, shape <span class="op">=</span> <span class="va">pop</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.2</span>, height <span class="op">=</span> <span class="fl">.2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'TRB'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'TRA'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span><span class="op">)</span><span class="op">)</span>

<span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_axis_labels.html">map_axis_labels</a></span><span class="op">(</span><span class="va">pairs_plt</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx1_tbl</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx2_tbl</span>, aes_label  <span class="op">=</span> <span class="st">'chain'</span><span class="op">)</span>
<span class="va">pairs_plt</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>By setting
<code>min_expansion = Inf, cluster_whitelist = whitelist</code> we can
examine any pairings for a set of cluster_idx, in this case the ones
that were seen multiple times. Interestingly (and unlike some human
samples) the expanded clusters are <span class="math inline">\(\beta\)</span>-chain, and their <span class="math inline">\(\alpha\)</span> chains are sprinkled quite evenly
across clusters.</p>
</div>
</div>
<div class="section level2">
<h2 id="permutation-tests">Permutation tests<a class="anchor" aria-label="anchor" href="#permutation-tests"></a>
</h2>
<p>Permutation tests allow tests of independence between cluster
assignments and other cell-level covariates (such as the sample from
which the cell was derived). The cluster label is permuted to break the
link between cell and cluster, and an arbitrary statistic of both
cluster label, and cell covariate is evaluated.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80_chain</span> <span class="op">=</span> <span class="fu"><a href="../reference/split_cdb.html">split_cdb</a></span><span class="op">(</span><span class="va">aa80</span>, <span class="st">'chain'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">canonicalize_cell</span>, contig_fields <span class="op">=</span> <span class="st">'aa80'</span><span class="op">)</span>

<span class="va">compare_expanded</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># cluster_idx contains the permuted cluster assignments</span>
  <span class="co"># grp the cell_covariate_keys.</span>
  <span class="co"># NB: this is always a data.frame even if it is just a single column</span>
  <span class="co"># cross tab by pop</span>
  <span class="va">tab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># count number of times an aa80 class was expanded</span>
  <span class="va">expanded</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">tab</span><span class="op">&gt;=</span><span class="fl">2</span><span class="op">)</span>
  <span class="co"># compare difference</span>
  <span class="va">expanded</span><span class="op">[</span><span class="st">'b6'</span><span class="op">]</span> <span class="op">-</span> <span class="va">expanded</span><span class="op">[</span><span class="st">'balbc'</span><span class="op">]</span>
<span class="op">}</span></code></pre></div>
<p>The signature of the statistic should be of a vector
<code>cluster_idx</code> and <code>data.frame</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">perm1</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>, cell_covariate_keys <span class="op">=</span> <span class="st">'pop'</span>, cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">100</span>, statistic <span class="op">=</span> <span class="va">compare_expanded</span><span class="op">)</span>

<span class="va">perm1</span>
<span class="co">#&gt; A single PermuteTest:</span>
<span class="co">#&gt; contrast identity. observed = 5, expected = 4.92, p.value = 0.86.</span></code></pre></div>
<p>Although b6 mice had 5 more clones observed to be expanded (occuring
&gt;=2 times) than balbc, this is not signficant under a null model
where cells were permuted between mouse types (populations), where b6
are expected to have about 5 more expanded clones, just due to the
additional number of cells sampled in b6 and the particular spectrum of
clonal frequencies in this experiment:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>pop <span class="op">=</span> <span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">pop</th>
<th align="right">Freq</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">b6</td>
<td align="right">398</td>
</tr>
<tr class="even">
<td align="left">balbc</td>
<td align="right">377</td>
</tr>
</tbody>
</table>
<p>Indeed if we resample in a way that fixes each group to have the same
number of cells:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rarify</span> <span class="op">=</span> <span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">cell_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">pop</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html" class="external-link">do</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_sample</a></span><span class="op">(</span><span class="va">.</span>, n <span class="op">=</span> <span class="fl">377</span><span class="op">)</span><span class="op">)</span>

<span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">cell_tbl</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">semi_join</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">cell_tbl</span>, <span class="va">rarify</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = c("aa80", "barcode", "pop", "sample")</span>

<span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>, cell_covariate_keys <span class="op">=</span> <span class="st">'pop'</span>, cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">500</span>, statistic <span class="op">=</span> <span class="va">compare_expanded</span><span class="op">)</span>
<span class="co">#&gt; A single PermuteTest:</span>
<span class="co">#&gt; contrast identity. observed = -1, expected = 0.726, p.value = 0.74.</span></code></pre></div>
<p>We see that this discrepacy between the number of expanded clones
between subpopulations is mostly explained by a greater number of cells
sampled in b6, but also random variability plays a role.</p>
<p>We can also test for oligoclonality, eg, how often is a beta chain
expanded in a sample:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_expanded</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># clusters x sample contigency table</span>
  <span class="va">tab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># number of cluster x samples that occured more than once</span>
  <span class="va">expanded</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tab</span><span class="op">&gt;</span><span class="fl">1</span><span class="op">)</span>
  <span class="va">expanded</span>
<span class="op">}</span>

<span class="va">perm3</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>,  cell_covariate_keys <span class="op">=</span> <span class="st">'sample'</span>, cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">500</span>, statistic <span class="op">=</span> <span class="va">count_expanded</span><span class="op">)</span>
<span class="va">perm3</span>
<span class="co">#&gt; A single PermuteTest:</span>
<span class="co">#&gt; contrast identity. observed = 27, expected = 37, p.value = 0.032.</span></code></pre></div>
<p>27 expanded clones were observed in each of the two populations vs 37
expected, and this discrepancy would be significant at <span class="math inline">\(p&lt;\)</span> 0.04. This is indicating that there
is underdispersion – fewer clusters are expanded than expected, given
the spectrum of clonal frequencies and the number of cells per
sample.</p>
<p>To further elucidate this, we can restrict the permutations to
maintain certain margins of the table by specifying
<code>cell_stratify_keys.</code> This doesn’t effect the observed values
of the statistics, but will change the expected values (since these are
now conditional expectations.) Here we restrict the permutations within
levels of <code>pop</code> (eg, only permuting within balbc, and within
b6).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>,   cell_covariate_keys <span class="op">=</span> <span class="st">'sample'</span>, cell_stratify_keys <span class="op">=</span> <span class="st">'pop'</span>, cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">500</span>, statistic <span class="op">=</span> <span class="va">count_expanded</span><span class="op">)</span>
<span class="co">#&gt; A single PermuteTest:</span>
<span class="co">#&gt; contrast identity. observed = 27, expected = 44.7, p.value = 0.002.</span></code></pre></div>
<p>In the restricted permutations, the expected number of expanded
clusters is even greater. Both of these effects are due to the fact that
the “sample” replicates, within each population actually are not
biological replicates, which inflates the <code>cluster_idx</code>
margin of the table.</p>
<div class="section level3">
<h3 id="sequences-of-permutation-tests-across-cell-subpopulations">Sequences of permutation tests across cell subpopulations<a class="anchor" aria-label="anchor" href="#sequences-of-permutation-tests-across-cell-subpopulations"></a>
</h3>
<p>In many cases, we want test a sequence of contrasts of
<code>cell_covariate_keys</code> variables vs the
<code>cell_label_key</code>. For instance,
<code>cell_covariate_keys</code> might include the cell subpopulation
derived from gene expression taken from a cell, and we want to compare
the levels of clonal expansion between subpopulations. Suppose we have
four such subpopulation <code>ident</code>s:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ident</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/gl.html" class="external-link">gl</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">cell_tbl</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">4</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ident</span><span class="op">)</span>
<span class="co">#&gt; [1] 1 2 3 4 1 2</span>
<span class="co">#&gt; Levels: 1 2 3 4</span>
<span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span><span class="op">$</span><span class="va">cell_tbl</span><span class="op">$</span><span class="va">ident</span> <span class="op">=</span> <span class="va">ident</span></code></pre></div>
<p>We can compare all pairs: 2 vs 1, 3 vs 1, 4 vs 1, 3 vs 2, 4 vs 2, 4
vs 3, and in fact this is the default action. If <code>contrasts</code>
(a matrix or list of vectors) is specified, we can perform other sets of
comparisons, like 1 vs the average in all other <code>ident</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">contrast_vec</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>The key is that the <code>statistic</code> should return a vector, as
this now does:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">compare_expanded_vec</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">)</span><span class="op">{</span>
  <span class="va">tab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cluster_idx</span>, <span class="va">grp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># count number of times an aa80 class was expanded</span>
  <span class="va">expanded</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">tab</span><span class="op">&gt;=</span><span class="fl">2</span><span class="op">)</span>
  <span class="va">expanded</span>
<span class="op">}</span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm4</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>, cell_covariate_keys <span class="op">=</span> <span class="st">'ident'</span>, cell_stratify_keys <span class="op">=</span> <span class="st">'sample'</span>, 
                             cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">100</span>, statistic <span class="op">=</span> <span class="va">compare_expanded_vec</span><span class="op">)</span>

<span class="fu"><a href="../reference/plot_permute_test.html">plot_permute_test</a></span><span class="op">(</span><span class="va">perm4</span><span class="op">)</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/perm-pairwise-1.png" width="700"></p>
<p>The default of all pairwise comparisons.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm5</span> <span class="op">=</span> <span class="fu"><a href="../reference/cluster_permute_test.html">cluster_permute_test</a></span><span class="op">(</span><span class="va">aa80_chain</span><span class="op">$</span><span class="va">TRB</span>, cell_covariate_keys <span class="op">=</span> <span class="st">'ident'</span>, cell_stratify_keys <span class="op">=</span> <span class="st">'sample'</span>, 
                             cell_label_key <span class="op">=</span> <span class="st">'aa80'</span>, n_perm <span class="op">=</span> <span class="fl">100</span>, statistic <span class="op">=</span> <span class="va">compare_expanded_vec</span>, contrasts <span class="op">=</span> <span class="va">contrast_vec</span><span class="op">)</span>

<span class="fu"><a href="../reference/plot_permute_test.html">plot_permute_test</a></span><span class="op">(</span><span class="va">perm5</span><span class="op">)</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/perm-avg-1.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_permute_test.html">tidy.PermuteTestList</a></span><span class="op">(</span><span class="va">perm5</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 100 × 7</span></span>
<span class="co">#&gt;    statistics observed expected p.value mc.se contrast_name contrast            </span>
<span class="co">#&gt;         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>  -<span style="color: #BB0000;">2.00</span><span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>  -<span style="color: #BB0000;">3</span>   <span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>  -<span style="color: #BB0000;">2.00</span><span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>   1.67<span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>   1.67<span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>  -<span style="color: #BB0000;">8.33</span><span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>  -<span style="color: #BB0000;">7</span>   <span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>  -<span style="color: #BB0000;">3.33</span><span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span>    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>  -<span style="color: #BB0000;">2.67</span><span style="color: #949494;">e</span>+ 0    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>   2.78<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span>    -<span style="color: #BB0000;">4.67</span>    0.677    0.12 0.346 1             1, -0.3333333333333…</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 90 more rows</span></span></code></pre></div>
<p>There is also a <code>tidy</code> method.</p>
</div>
</div>
<div class="section level2">
<h2 id="colophone">Colophone<a class="anchor" aria-label="anchor" href="#colophone"></a>
</h2>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Monterey 12.3.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggdendro_0.1.23        purrr_0.3.4            stringr_1.4.0         </span>
<span class="co">#&gt; [4] tidyr_1.2.0            readr_2.1.2            ggplot2_3.3.5         </span>
<span class="co">#&gt; [7] dplyr_1.0.8            CellaRepertorium_1.7.1 BiocStyle_2.22.0      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] nlme_3.1-153           bitops_1.0-7           fs_1.5.2              </span>
<span class="co">#&gt;  [4] RColorBrewer_1.1-2     progress_1.2.2         rprojroot_2.0.2       </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.30.1    tools_4.1.2            backports_1.4.1       </span>
<span class="co">#&gt; [10] bslib_0.3.1            utf8_1.2.2             R6_2.5.1              </span>
<span class="co">#&gt; [13] DBI_1.1.2              BiocGenerics_0.40.0    colorspace_2.0-3      </span>
<span class="co">#&gt; [16] withr_2.5.0            prettyunits_1.1.1      tidyselect_1.1.2      </span>
<span class="co">#&gt; [19] compiler_4.1.2         textshaping_0.3.6      cli_3.2.0             </span>
<span class="co">#&gt; [22] desc_1.4.1             labeling_0.4.2         bookdown_0.25         </span>
<span class="co">#&gt; [25] sass_0.4.1             scales_1.1.1           pkgdown_2.0.2         </span>
<span class="co">#&gt; [28] systemfonts_1.0.4      digest_0.6.29          minqa_1.2.4           </span>
<span class="co">#&gt; [31] rmarkdown_2.13         XVector_0.34.0         pkgconfig_2.0.3       </span>
<span class="co">#&gt; [34] htmltools_0.5.2        lme4_1.1-28            fastmap_1.1.0         </span>
<span class="co">#&gt; [37] highr_0.9              rlang_1.0.2            rstudioapi_0.13       </span>
<span class="co">#&gt; [40] jquerylib_0.1.4        farver_2.1.0           generics_0.1.2        </span>
<span class="co">#&gt; [43] jsonlite_1.8.0         broom.mixed_0.2.7      RCurl_1.98-1.6        </span>
<span class="co">#&gt; [46] magrittr_2.0.2         GenomeInfoDbData_1.2.7 Matrix_1.3-4          </span>
<span class="co">#&gt; [49] Rcpp_1.0.8             munsell_0.5.0          S4Vectors_0.32.3      </span>
<span class="co">#&gt; [52] fansi_1.0.2            lifecycle_1.0.1        stringi_1.7.6         </span>
<span class="co">#&gt; [55] yaml_2.3.5             MASS_7.3-54            zlibbioc_1.40.0       </span>
<span class="co">#&gt; [58] plyr_1.8.6             grid_4.1.2             forcats_0.5.1         </span>
<span class="co">#&gt; [61] crayon_1.5.0           lattice_0.20-45        Biostrings_2.62.0     </span>
<span class="co">#&gt; [64] cowplot_1.1.1          splines_4.1.2          hms_1.1.1             </span>
<span class="co">#&gt; [67] knitr_1.37             pillar_1.7.0           boot_1.3-28           </span>
<span class="co">#&gt; [70] reshape2_1.4.4         stats4_4.1.2           glue_1.6.2            </span>
<span class="co">#&gt; [73] evaluate_0.15          BiocManager_1.30.16    vctrs_0.3.8           </span>
<span class="co">#&gt; [76] nloptr_2.0.0           tzdb_0.2.0             gtable_0.3.0          </span>
<span class="co">#&gt; [79] assertthat_0.2.1       cachem_1.0.6           xfun_0.30             </span>
<span class="co">#&gt; [82] broom_0.7.12           ragg_1.2.2             tibble_3.1.6          </span>
<span class="co">#&gt; [85] memoise_2.0.1          IRanges_2.28.0         ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Andrew McDavid, Yu Gu, Erik VonKaenel, Aaron Wagner.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
