<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering repertoire via CDR3 sequences • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering repertoire via CDR3 sequences">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering repertoire via CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Clustering repertoire via CDR3 sequences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/master/vignettes/cdr3_clustering.Rmd"><code>vignettes/cdr3_clustering.Rmd</code></a></small>
      <div class="hidden name"><code>cdr3_clustering.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(CellaRepertorium)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; Attaching package: 'dplyr'</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt;     filter, lag</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt;     intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(readr)</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyr)</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(stringr)</a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(purrr)</a></code></pre></div>
<div id="load-filtered-contig-files" class="section level1">
<h1 class="hasAnchor">
<a href="#load-filtered-contig-files" class="anchor"></a>Load filtered contig files</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(contigs_qc)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">MIN_CDR3_AA =<span class="st"> </span><span class="dv">6</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5">cdb =<span class="st"> </span><span class="kw"><a href="../reference/ContigCellDB-fun.html">ContigCellDB_10XVDJ</a></span>(contigs_qc, <span class="dt">contig_pk =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>, <span class="st">'contig_id'</span>), <span class="dt">cell_pk =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>))</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">cdb<span class="op">$</span>contig_tbl =<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(cdb<span class="op">$</span>contig_tbl, full_length, productive <span class="op">==</span><span class="st"> 'True'</span>, high_confidence, chain <span class="op">!=</span><span class="st"> 'Multi'</span>, <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(cdr3) <span class="op">&gt;</span><span class="st"> </span>MIN_CDR3_AA) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>( <span class="dt">fancy_name =</span> <span class="kw"><a href="../reference/fancy_name_contigs.html">fancy_name_contigs</a></span>(., <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_c.html">str_c</a></span>(pop, <span class="st">'_'</span>, sample)))</a></code></pre></div>
<p>good chains (either TRA or TRB); each cell can appear more than once.</p>
</div>
<div id="chain-pairings" class="section level1">
<h1 class="hasAnchor">
<a href="#chain-pairings" class="anchor"></a>Chain pairings</h1>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">paired_chain =<span class="st"> </span><span class="kw"><a href="../reference/enumerate_pairing.html">enumerate_pairing</a></span>(cdb, <span class="dt">chain_recode_fun =</span> <span class="st">'guess'</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(paired_chain, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/interaction">interaction</a></span>(sample, pop), <span class="dt">fill =</span> pairing)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>canonical, <span class="dt">scale =</span> <span class="st">'free_x'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div id="cluster-cdr3-protein-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-protein-sequences" class="anchor"></a>Cluster CDR3 protein sequences</h1>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">aa80 =<span class="st"> </span><span class="kw"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(cdb, <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">cluster_name =</span> <span class="st">'aa80'</span>, <span class="dt">identity =</span> <span class="fl">.8</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">aa80 =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(aa80, <span class="dt">sequence_key =</span> <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">keep_clustering_details =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt; Calculating intradistances on 988 clusters.</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; Summarizing</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co"># This maybe should be a turned into a function </span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co"># Other plots should be considered:</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co"># That show how clusters are split between samples, chains, etc</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(aa80<span class="op">$</span>cluster_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(n_cluster<span class="op">&gt;</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(key, value, <span class="op">-</span>aa80, <span class="op">-</span>fc) , <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> value))<span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>key, <span class="dt">scales =</span> <span class="st">'free'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()</a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We cluster the CDR3 translated amino acid residues with the program <a href="http://weizhongli-lab.org/cdhit_suite/cgi-bin/index.cgi?cmd=cd-hit">CD-HIT</a>. A sequence is included in a cluster if it matches by 100% similarity and has the same CDR3 length. Note that this can and should be relaxed – especially in the beta chain we see “near clones” that only differ by a residue or two, seemingly in stylized places.</p>
</div>
<div id="cluster-cdr3-dna-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-dna-sequences" class="anchor"></a>Cluster CDR3 DNA sequences</h1>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">cdb =<span class="st"> </span><span class="kw"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(cdb, <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>, <span class="dt">cluster_name =</span> <span class="st">'DNA97'</span>, <span class="dt">identity =</span> <span class="fl">.965</span>, <span class="dt">min_length =</span> MIN_CDR3_AA<span class="op">*</span><span class="dv">3-1</span>, <span class="dt">G =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">cdb =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(cdb, <span class="dt">sequence_key =</span> <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; Calculating intradistances on 1342 clusters.</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; Summarizing</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(cdb<span class="op">$</span>cluster_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(n_cluster<span class="op">&gt;</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(key, value, <span class="op">-</span>DNA97) , <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> value))<span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>key, <span class="dt">scales =</span> <span class="st">'free'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We can also cluster by DNA identity.</p>
</div>
<div id="cluster-by-v-j-identity" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-by-v-j-identity" class="anchor"></a>Cluster by V-J identity</h1>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">germline_cluster =<span class="st"> </span><span class="kw"><a href="../reference/cluster_germline.html">cluster_germline</a></span>(cdb, <span class="dt">segment_keys =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'chain'</span>), <span class="dt">cluster_name =</span> <span class="st">'segment_idx'</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; Warning in replace_cluster_tbl(ccdb, cluster_tbl, cl_con_tbl, cluster_pk =</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; cluster_name): Replacing `cluster_tbl` with DNA97, .</span></a></code></pre></div>
<p>We can cluster by any other feature of the contigs. Here we cluster each contig based on the chain and V-J genes. This gives us the set of observed V-J pairings:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">germline_cluster =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(germline_cluster, <span class="dt">sequence_key =</span> <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">#&gt; Calculating intradistances on 700 clusters.</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#&gt; Summarizing</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt; Warning in left_join_warn(d_medoid, contig_tbl, by = ccdb$contig_pk,</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt; overwrite = TRUE): Overwriting fields d(medoid), is_medoid in table</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt; contig_tbl</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(germline_cluster<span class="op">$</span>cluster_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(chain <span class="op">==</span><span class="st"> 'TRB'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> v_gene, <span class="dt">y =</span> j_gene, <span class="dt">fill =</span> n_cluster)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>))</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Number of pairs</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(germline_cluster<span class="op">$</span>cluster_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(chain <span class="op">==</span><span class="st"> 'TRB'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> v_gene, <span class="dt">y =</span> j_gene, <span class="dt">fill =</span> avg_distance)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>))</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Average Levenshtein distance of CDR3 within each pair. This might be turned into a z-score by fitting a weighted linear model with sum-to-zero contrasts and returning the pearson residuals. This could determine if a pairing has an unexpected small, or large, within cluster distance.</p>
</div>
<div id="oligo-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#oligo-clusters" class="anchor"></a>Oligo clusters</h1>
<p>Next, we will examine the clusters that are found in many contigs. First we will get a canonical contig to represent each cluster. This will be the medoid contig, by default.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">aa80 =<span class="st"> </span><span class="kw"><a href="../reference/canonicalize_cluster.html">canonicalize_cluster</a></span>(aa80, <span class="dt">representative =</span> <span class="st">'cdr3'</span>, <span class="dt">contig_fields =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'cdr3_nt'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'d_gene'</span>, <span class="st">'j_gene'</span>))</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co">#&gt; Filtering `contig_tbl` by `is_medoid`, override by setting `contig_filter_args == TRUE`</span></a></code></pre></div>
<p><code>aa80</code> now includes the fields listed in <code>contig_fields</code> in the <code>cluster_tbl</code>, using the values found in the medoid contig.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">MIN_OLIGO =<span class="st"> </span><span class="dv">7</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">oligo_clusters =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(aa80<span class="op">$</span>cluster_tbl, n_cluster <span class="op">&gt;=</span><span class="st"> </span>MIN_OLIGO)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">oligo_contigs =<span class="st"> </span>aa80</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">oligo_contigs<span class="op">$</span>contig_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">semi_join</a></span>(oligo_contigs<span class="op">$</span>contig_tbl, oligo_clusters, <span class="dt">by =</span> <span class="st">'aa80'</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">oligo_contigs</a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co">#&gt; ContigCellDB of 54 contigs; 54 cells; and 4 clusters.</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">#&gt; Contigs keyed by barcode, pop, sample, contig_id; cells keyed by barcode, pop, sample.</span></a></code></pre></div>
<p>Get contigs/cells/clusters found at least 7 times (across contigs). Note that replacing <code>contig_tbl</code> with the subset selected with the <code>semi_join</code> also automatically subsetted the <code>cell_tbl</code> and <code>cluster_tbl</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">oligo_clusters =<span class="st"> </span>oligo_contigs<span class="op">$</span>contig_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(aa80) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="st">`</span><span class="dt">n subjects observed</span><span class="st">`</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(sample))) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(oligo_clusters)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#&gt; Joining, by = "aa80"</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"></a>
<a class="sourceLine" id="cb11-4" data-line-number="4">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(oligo_clusters <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(aa80<span class="op">:</span>cdr3, chain<span class="op">:</span>j_gene, avg_distance, n_cluster))</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">aa80</th>
<th align="right">n subjects observed</th>
<th align="left">cdr3</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="right">avg_distance</th>
<th align="right">n_cluster</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">111</td>
<td align="right">6</td>
<td align="left">CVVGDRGSALGRLHF</td>
<td align="left">TRA</td>
<td align="left">TRAV11</td>
<td align="left">None</td>
<td align="left">TRAJ18</td>
<td align="right">0.6071429</td>
<td align="right">28</td>
</tr>
<tr class="even">
<td align="right">172</td>
<td align="right">5</td>
<td align="left">CAVSRASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV9N-3</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">2.1111111</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="right">296</td>
<td align="right">6</td>
<td align="left">CAASASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV14D-2</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">1.5000000</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="right">808</td>
<td align="right">4</td>
<td align="left">CATGNYAQGLTF</td>
<td align="left">TRA</td>
<td align="left">TRAV8D-2</td>
<td align="left">None</td>
<td align="left">TRAJ26</td>
<td align="right">1.3333333</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p>Report some statistics about these expanded clusters.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">oligo_plot =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(oligo_contigs<span class="op">$</span>contig_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> representative, <span class="dt">fill =</span> chain)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">type =</span> <span class="st">'qual'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">oligo_plot</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>These always come from a single chain.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">oligo_plot <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">fill =</span>   sample) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>pop)</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>But come from multiple populations and samples.</p>
<div id="some-simple-phylogenetic-relationships" class="section level2">
<h2 class="hasAnchor">
<a href="#some-simple-phylogenetic-relationships" class="anchor"></a>Some simple phylogenetic relationships</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggdendro)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="co"># This should be turned into a function in the package somehow</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="co"># But plot arguments will be super-variable</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="co"># Maybe just return the `hc` object?</span></a>
<a class="sourceLine" id="cb14-6" data-line-number="6">dendro_plot =<span class="st"> </span><span class="cf">function</span>(ccdb, idx, <span class="dt">method =</span> <span class="st">'complete'</span>){</a>
<a class="sourceLine" id="cb14-7" data-line-number="7">    h =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(ccdb<span class="op">$</span>cluster_tbl, <span class="op">!!</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(ccdb<span class="op">$</span>cluster_pk) <span class="op">==</span><span class="st"> </span>idx) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span>(fc) <span class="op">%&gt;%</span><span class="st"> </span>.[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb14-8" data-line-number="8">    quer =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(ccdb<span class="op">$</span>contig_tbl, <span class="op">!!</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(ccdb<span class="op">$</span>cluster_pk) <span class="op">==</span><span class="st"> </span>idx)</a>
<a class="sourceLine" id="cb14-9" data-line-number="9">    hc =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/hclust">hclust</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/dist">as.dist</a></span>(h<span class="op">$</span>distance_mat), <span class="dt">method =</span> method) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggdendro/topics/dendro_data">dendro_data</a></span>(<span class="dt">type =</span> <span class="st">"rectangle"</span>)</a>
<a class="sourceLine" id="cb14-10" data-line-number="10">    hc<span class="op">$</span>labels =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(hc<span class="op">$</span>labels, quer)</a>
<a class="sourceLine" id="cb14-11" data-line-number="11">   <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(hc<span class="op">$</span>segments, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>x, <span class="dt">y=</span>y)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">xend=</span>xend, <span class="dt">yend=</span>yend)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-12" data-line-number="12"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span>(<span class="dt">data =</span> hc<span class="op">$</span>labels, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">color =</span> sample, <span class="dt">label =</span> fancy_name), <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">angle =</span> <span class="dv">60</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">breaks =</span> <span class="ot">NULL</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'AA Distance'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">''</span>)</a>
<a class="sourceLine" id="cb14-13" data-line-number="13">}</a>
<a class="sourceLine" id="cb14-14" data-line-number="14"></a>
<a class="sourceLine" id="cb14-15" data-line-number="15">to_plot =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(aa80<span class="op">$</span>cluster_tbl, n_cluster <span class="op">&gt;=</span><span class="st"> </span>MIN_OLIGO)</a>
<a class="sourceLine" id="cb14-16" data-line-number="16"></a>
<a class="sourceLine" id="cb14-17" data-line-number="17"><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(to_plot<span class="op">$</span>aa80, <span class="op">~</span><span class="st"> </span><span class="kw">dendro_plot</span>(aa80, .))</a>
<a class="sourceLine" id="cb14-18" data-line-number="18"><span class="co">#&gt; [[1]]</span></a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-2.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-3.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-4.png" width="700"></p>
</div>
<div id="formal-testing-for-frequency-differences" class="section level2">
<h2 class="hasAnchor">
<a href="#formal-testing-for-frequency-differences" class="anchor"></a>Formal testing for frequency differences</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(lme4)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(broom)</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">per_chain_sample =<span class="st"> </span>good_cluster_cells <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(sample, pop, chain) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="dt">total_cells =</span> <span class="kw"><a href="https://dplyr.tidyverse.org/reference/n.html">n</a></span>(), <span class="dt">weeks_premature =</span> weeks_premature[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb18-4" data-line-number="4"></a>
<a class="sourceLine" id="cb18-5" data-line-number="5">oligo_cluster_stat =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">semi_join</a></span>(oligo_clusters, good_cluster_cells <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(dataset, contig_id)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(sample, pop, chain, cluster_idx) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="dt">n_cluster =</span> <span class="kw"><a href="https://dplyr.tidyverse.org/reference/n.html">n</a></span>())<span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/complete.html">complete</a></span>(sample, pop, <span class="kw"><a href="https://tidyr.tidyverse.org/reference/expand.html">nesting</a></span>(cluster_idx, chain), <span class="dt">fill =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">n_cluster =</span> <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb18-6" data-line-number="6"></a>
<a class="sourceLine" id="cb18-7" data-line-number="7">oligo_cluster_stat =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(oligo_cluster_stat, per_chain_sample, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'sample'</span>, <span class="st">'pop'</span>, <span class="st">'chain'</span>)) </a>
<a class="sourceLine" id="cb18-8" data-line-number="8"></a>
<a class="sourceLine" id="cb18-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stopifnot">stopifnot</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all">all</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colSums</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/with">with</a></span>(oligo_cluster_stat, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(chain, cluster_idx)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>))</a>
<a class="sourceLine" id="cb18-10" data-line-number="10"></a>
<a class="sourceLine" id="cb18-11" data-line-number="11">mm_out =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(oligo_cluster_stat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(cluster_idx, chain) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/lme4/topics/glmer">glmer</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(n_cluster, total_cells) <span class="op">~</span><span class="st"> </span>pop <span class="op">+</span><span class="st"> </span>weeks_premature <span class="op">+</span><span class="st"> </span>(<span class="dv">1</span><span class="op">|</span>sample), <span class="dt">data =</span> ., <span class="dt">family =</span> <span class="st">'binomial'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/broom/topics/reexports">tidy</a></span>(<span class="dt">conf.int =</span> <span class="ot">TRUE</span>)))</a></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">mm_outj =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>(mm_out), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(oligo_clusters_all <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(representative, cluster_idx))), term <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'popCD31Pos'</span>, <span class="st">'weeks_premature'</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">ci_lo =</span> AMmisc<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/AMmisc/topics/clamp">clamp</a></span>(conf.low), <span class="dt">ci_hi =</span> AMmisc<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/AMmisc/topics/clamp">clamp</a></span>(conf.high)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(representative))</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(mm_outj, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> representative, <span class="dt">ymin =</span> ci_lo, <span class="dt">ymax =</span> ci_hi, <span class="dt">y =</span> <span class="kw">clamp</span>(estimate))) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>term, <span class="dt">scales =</span> <span class="st">'free'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">lty =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Isomorph"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"log odds of isomorph"</span>)</a></code></pre></div>
<p>We test if the binomial rate of clone expression differs between CD31+/- or term, for each clone.</p>
</div>
</div>
<div id="clonal-pairs" class="section level1">
<h1 class="hasAnchor">
<a href="#clonal-pairs" class="anchor"></a>Clonal pairs</h1>
<p>Next, we can examine the pairing between <span class="math inline">\(\alpha-\beta\)</span> chains and see if any pairs are found more than once.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">class_colors =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">tibble</a></span>(<span class="dt">chain =</span>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(aa80<span class="op">$</span>cluster_tbl<span class="op">$</span>chain)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span>  RColorBrewer<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer">brewer.pal</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(chain),<span class="st">"Set1"</span>)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_along</a></span>(chain)])</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="co">#&gt; Warning in RColorBrewer::brewer.pal(length(chain), "Set1"): minimal value for n is 3, returning requested palette with 3 different levels</span></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"></a>
<a class="sourceLine" id="cb20-4" data-line-number="4">aa80<span class="op">$</span>cluster_pk =<span class="st"> 'representative'</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5">pairing_list =<span class="st"> </span><span class="kw"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(aa80, <span class="dt">table_order =</span> <span class="dv">2</span>, <span class="dt">orphan_level =</span> <span class="dv">1</span>, <span class="dt">min_expansion =</span> <span class="dv">3</span>, <span class="dt">cluster_keys =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))</a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="co">#&gt; Warning: Factor `cluster_idx.2` contains implicit NA, consider using</span></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="co">#&gt; `forcats::fct_explicit_na`</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8"><span class="co">#&gt; Warning: Column `representative` joining factors with different levels,</span></a>
<a class="sourceLine" id="cb20-9" data-line-number="9"><span class="co">#&gt; coercing to character vector</span></a></code></pre></div>
<p><code>pairing_tables</code> finds all contig combinations of order <code>table_order</code> across cells. Among those combinations that occur at least <code>min_expansion</code> times, the expanded combinations and and any other combinations that shared an expanded combo.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">pairs_plt =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pairing_list<span class="op">$</span>cell_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> cluster_idx<span class="fl">.1</span>_fct, <span class="dt">y =</span> cluster_idx<span class="fl">.2</span>_fct, <span class="dt">color =</span> sample, <span class="dt">shape =</span> pop)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="dt">width =</span> <span class="fl">.2</span>, <span class="dt">height =</span> <span class="fl">.2</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"></a>
<a class="sourceLine" id="cb21-3" data-line-number="3">feature_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(pairing_list<span class="op">$</span>idx1_tbl, pairing_list<span class="op">$</span>idx2_tbl), class_colors)</a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,</span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector,</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="co">#&gt; Joining, by = "chain"</span></a>
<a class="sourceLine" id="cb21-9" data-line-number="9"></a>
<a class="sourceLine" id="cb21-10" data-line-number="10">ylab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)<span class="op">$</span>layout<span class="op">$</span>panel_params[[<span class="dv">1</span>]]<span class="op">$</span>y.label) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))</a>
<a class="sourceLine" id="cb21-11" data-line-number="11"><span class="co">#&gt; Warning: `data_frame()` is deprecated, use `tibble()`.</span></a>
<a class="sourceLine" id="cb21-12" data-line-number="12"><span class="co">#&gt; This warning is displayed once per session.</span></a>
<a class="sourceLine" id="cb21-13" data-line-number="13"><span class="co">#&gt; Joining, by = "representative"</span></a>
<a class="sourceLine" id="cb21-14" data-line-number="14"></a>
<a class="sourceLine" id="cb21-15" data-line-number="15">xlab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)<span class="op">$</span>layout<span class="op">$</span>panel_params[[<span class="dv">1</span>]]<span class="op">$</span>x.label) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))</a>
<a class="sourceLine" id="cb21-16" data-line-number="16"><span class="co">#&gt; Joining, by = "representative"</span></a>
<a class="sourceLine" id="cb21-17" data-line-number="17"></a>
<a class="sourceLine" id="cb21-18" data-line-number="18">pairs_plt =<span class="st"> </span>pairs_plt <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">color =</span> xlab<span class="op">$</span>class_color, <span class="dt">size =</span> <span class="dv">8</span>), <span class="dt">axis.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">color =</span> ylab<span class="op">$</span>class_color, <span class="dt">size =</span> <span class="dv">8</span>))</a>
<a class="sourceLine" id="cb21-19" data-line-number="19"></a>
<a class="sourceLine" id="cb21-20" data-line-number="20">pairs_plt</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/plot_expanded-1.png" width="700"></p>
<div id="expanded-clones" class="section level2">
<h2 class="hasAnchor">
<a href="#expanded-clones" class="anchor"></a>Expanded clones</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">whitelist =<span class="st"> </span>oligo_clusters <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="dt">cluster_idx.1 =</span> representative) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>()</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">pairing_list =<span class="st"> </span><span class="kw"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(aa80, <span class="dt">canonicalize_fun =</span> canonicalize_by_prevalence, <span class="dt">table_order =</span> <span class="dv">2</span>, <span class="dt">orphan_level =</span> <span class="dv">1</span>, <span class="dt">min_expansion =</span> <span class="ot">Inf</span>, <span class="dt">cluster_whitelist =</span> whitelist,  <span class="dt">cluster_keys =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">pairs_plt =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pairing_list<span class="op">$</span>cell_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> cluster_idx<span class="fl">.1</span>_fct, <span class="dt">y =</span> cluster_idx<span class="fl">.2</span>_fct, <span class="dt">color =</span> sample, <span class="dt">shape =</span> pop)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="dt">width =</span> <span class="fl">.2</span>, <span class="dt">height =</span> <span class="fl">.2</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>)</a>
<a class="sourceLine" id="cb22-4" data-line-number="4"></a>
<a class="sourceLine" id="cb22-5" data-line-number="5">feature_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(pairing_list<span class="op">$</span>idx1_tbl, pairing_list<span class="op">$</span>idx2_tbl), class_colors)</a>
<a class="sourceLine" id="cb22-6" data-line-number="6"></a>
<a class="sourceLine" id="cb22-7" data-line-number="7">ylab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)<span class="op">$</span>layout<span class="op">$</span>panel_params[[<span class="dv">1</span>]]<span class="op">$</span>y.label) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))</a>
<a class="sourceLine" id="cb22-8" data-line-number="8"></a>
<a class="sourceLine" id="cb22-9" data-line-number="9">xlab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)<span class="op">$</span>layout<span class="op">$</span>panel_params[[<span class="dv">1</span>]]<span class="op">$</span>x.label) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))</a>
<a class="sourceLine" id="cb22-10" data-line-number="10"></a>
<a class="sourceLine" id="cb22-11" data-line-number="11">pairs_plt =<span class="st"> </span>pairs_plt <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">color =</span> xlab<span class="op">$</span>class_color, <span class="dt">size =</span> <span class="dv">8</span>), <span class="dt">axis.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">color =</span> ylab<span class="op">$</span>class_color, <span class="dt">size =</span> <span class="dv">8</span>))</a>
<a class="sourceLine" id="cb22-12" data-line-number="12"></a>
<a class="sourceLine" id="cb22-13" data-line-number="13">pairs_plt</a></code></pre></div>
<p>By setting <code>min_expansion = Inf, cluster_whitelist = whitelist</code> we can examine any pairings for a set of cluster_idx, in this case the ones that were seen multiple times. Interestingly (and unlike some human samples) the expanded clusters are <span class="math inline">\(\beta\)</span>-chain, and their <span class="math inline">\(\alpha\)</span> chains are sprinkled quite evenly across clusters.</p>
</div>
</div>
<div id="length-of-cdr3" class="section level1">
<h1 class="hasAnchor">
<a href="#length-of-cdr3" class="anchor"></a>Length of CDR3</h1>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">aa80<span class="op">$</span>contig_tbl =<span class="st"> </span>aa80<span class="op">$</span>contig_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">cdr3_length =</span> <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(cdr3_nt))</a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(aa80<span class="op">$</span>contig_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">fill =</span> pop, <span class="dt">x=</span> cdr3_length)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">binwidth =</span> <span class="dv">1</span>, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> ..density..)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">type =</span> <span class="st">'qual'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(sample <span class="op">~</span>chain) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">strip.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">xlim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">25</span>, <span class="dv">55</span>))</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Plot the CDR3 length distribution for each sample and pop. There doesn’t appear to be a noticeable difference between BALB/c and C57BL/6J (b6) mice, but if we needed to make sure, an appropriate procedure would be to run a mixed model with a random <code>sample</code> effect (assumed to represent a biological replicate).</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">cdr_len =<span class="st"> </span>aa80<span class="op">$</span>contig_tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(chain) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span>(broom<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/broom/topics/reexports">tidy</a></span>(lme4<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/lme4/topics/lmer">lmer</a></span>(cdr3_length <span class="op">~</span><span class="st"> </span>pop <span class="op">+</span><span class="st"> </span>(<span class="dv">1</span><span class="op">|</span>sample), <span class="dt">data =</span> .), <span class="dt">conf.int =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co">#&gt; boundary (singular) fit: see ?isSingular</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,</span></a>
<a class="sourceLine" id="cb24-4" data-line-number="4"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb24-5" data-line-number="5"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector,</span></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb24-7" data-line-number="7"><span class="co">#&gt; boundary (singular) fit: see ?isSingular</span></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,</span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb24-10" data-line-number="10"></a>
<a class="sourceLine" id="cb24-11" data-line-number="11"><span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector,</span></a>
<a class="sourceLine" id="cb24-12" data-line-number="12"><span class="co">#&gt; coercing into character vector</span></a>
<a class="sourceLine" id="cb24-13" data-line-number="13"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(cdr_len <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(group <span class="op">==</span><span class="st"> 'fixed'</span>, term <span class="op">!=</span><span class="st"> '(Intercept)'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/interaction">interaction</a></span>(chain, term), <span class="dt">y =</span> estimate, <span class="dt">ymin =</span> conf.low, <span class="dt">ymax =</span> conf.high)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'Length(CDR3 Nt)'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'Term/Chain'</span>)</a></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/cdr3_len-1.png" width="288"></p>
<p>We end up with a (harmless) convergence warning about a singular fit. This is expected, because the <code>samples</code> aren’t actually replicates – they are just subsamples drawn for illustrative purposes. The Balbc mice have .5 fewer nucleotides per contig, on average, and this is not significant.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-filtered-contig-files">Load filtered contig files</a></li>
      <li><a href="#chain-pairings">Chain pairings</a></li>
      <li><a href="#cluster-cdr3-protein-sequences">Cluster CDR3 protein sequences</a></li>
      <li><a href="#cluster-cdr3-dna-sequences">Cluster CDR3 DNA sequences</a></li>
      <li><a href="#cluster-by-v-j-identity">Cluster by V-J identity</a></li>
      <li>
<a href="#oligo-clusters">Oligo clusters</a><ul class="nav nav-pills nav-stacked">
<li><a href="#some-simple-phylogenetic-relationships">Some simple phylogenetic relationships</a></li>
      <li><a href="#formal-testing-for-frequency-differences">Formal testing for frequency differences</a></li>
      </ul>
</li>
      <li>
<a href="#clonal-pairs">Clonal pairs</a><ul class="nav nav-pills nav-stacked">
<li><a href="#expanded-clones">Expanded clones</a></li>
      </ul>
</li>
      <li><a href="#length-of-cdr3">Length of CDR3</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Yu Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
