<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering repertoire via CDR3 sequences • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering repertoire via CDR3 sequences">
<meta property="og:description" content="CellaRepertorium">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.98.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering repertoire via CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Clustering repertoire via CDR3 sequences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/master/vignettes/cdr3_clustering.Rmd"><code>vignettes/cdr3_clustering.Rmd</code></a></small>
      <div class="hidden name"><code>cdr3_clustering.Rmd</code></div>

    </div>

    
    
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">CellaRepertorium</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">readr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">stringr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">purrr</span>)</pre></body></html></div>
<div id="load-filtered-contig-files" class="section level1">
<h1 class="hasAnchor">
<a href="#load-filtered-contig-files" class="anchor"></a>Load filtered contig files</h1>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">contigs_qc</span>)
<span class="no">MIN_CDR3_AA</span> <span class="kw">=</span> <span class="fl">6</span>


<span class="no">cdb</span> <span class="kw">=</span> <span class="fu"><a href="../reference/ContigCellDB-fun.html">ContigCellDB_10XVDJ</a></span>(<span class="no">contigs_qc</span>, <span class="kw">contig_pk</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>, <span class="st">'contig_id'</span>), <span class="kw">cell_pk</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>))

<span class="no">cdb</span>$<span class="no">contig_tbl</span> <span class="kw">=</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">cdb</span>$<span class="no">contig_tbl</span>, <span class="no">full_length</span>, <span class="no">productive</span> <span class="kw">==</span> <span class="st">'True'</span>, <span class="no">high_confidence</span>, <span class="no">chain</span> <span class="kw">!=</span> <span class="st">'Multi'</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(<span class="no">cdr3</span>) <span class="kw">&gt;</span> <span class="no">MIN_CDR3_AA</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>( <span class="kw">fancy_name</span> <span class="kw">=</span> <span class="fu"><a href="../reference/fancy_name_contigs.html">fancy_name_contigs</a></span>(<span class="no">.</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html">str_c</a></span>(<span class="no">pop</span>, <span class="st">'_'</span>, <span class="no">sample</span>)))</pre></body></html></div>
<p>832 good chains (either TRA or TRB); each cell can appear more than once.</p>
</div>
<div id="chain-pairings" class="section level1">
<h1 class="hasAnchor">
<a href="#chain-pairings" class="anchor"></a>Chain pairings</h1>
<div class="sourceCode"><html><body><pre class="r"><span class="no">paired_chain</span> <span class="kw">=</span> <span class="fu"><a href="../reference/enumerate_pairing.html">enumerate_pairing</a></span>(<span class="no">cdb</span>, <span class="kw">chain_recode_fun</span> <span class="kw">=</span> <span class="st">'guess'</span>)

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">paired_chain</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span>(<span class="no">sample</span>, <span class="no">pop</span>), <span class="kw">fill</span> <span class="kw">=</span> <span class="no">pairing</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">canonical</span>, <span class="kw">scale</span> <span class="kw">=</span> <span class="st">'free_x'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div id="cluster-cdr3-protein-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-protein-sequences" class="anchor"></a>Cluster CDR3 protein sequences</h1>
<div class="sourceCode"><html><body><pre class="r">
<span class="no">aa80</span> <span class="kw">=</span> <span class="fu"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(<span class="no">cdb</span>, <span class="st">'cdr3'</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'AA'</span>, <span class="kw">cluster_name</span> <span class="kw">=</span> <span class="st">'aa80'</span>, <span class="kw">identity</span> <span class="kw">=</span> <span class="fl">.8</span>)
<span class="no">aa80</span> <span class="kw">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(<span class="no">aa80</span>, <span class="kw">sequence_key</span> <span class="kw">=</span> <span class="st">'cdr3'</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'AA'</span>, <span class="kw">keep_clustering_details</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; Calculating intradistances on 988 clusters.</span>
<span class="co">#&gt; Summarizing</span>


<span class="co"># This maybe should be a turned into a function </span>
<span class="co"># Other plots should be considered:</span>
<span class="co"># That show how clusters are split between samples, chains, etc</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">aa80</span>$<span class="no">cluster_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">n_cluster</span><span class="kw">&gt;</span><span class="fl">1</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(<span class="no">key</span>, <span class="no">value</span>, -<span class="no">aa80</span>, -<span class="no">fc</span>) , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">value</span>))+ <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">key</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">'free'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We cluster the CDR3 translated amino acid residues with the program <a href="http://weizhongli-lab.org/cdhit_suite/cgi-bin/index.cgi?cmd=cd-hit">CD-HIT</a>. A sequence is included in a cluster if it matches by 100% similarity and has the same CDR3 length. Note that this can and should be relaxed – especially in the beta chain we see “near clones” that only differ by a residue or two, seemingly in stylized places.</p>
</div>
<div id="cluster-cdr3-dna-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-dna-sequences" class="anchor"></a>Cluster CDR3 DNA sequences</h1>
<div class="sourceCode"><html><body><pre class="r"><span class="no">cdb</span> <span class="kw">=</span> <span class="fu"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(<span class="no">cdb</span>, <span class="st">'cdr3_nt'</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'DNA'</span>, <span class="kw">cluster_name</span> <span class="kw">=</span> <span class="st">'DNA97'</span>, <span class="kw">identity</span> <span class="kw">=</span> <span class="fl">.965</span>, <span class="kw">min_length</span> <span class="kw">=</span> <span class="no">MIN_CDR3_AA</span>*<span class="fl">3</span>-<span class="fl">1</span>, <span class="kw">G</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="no">cdb</span> <span class="kw">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(<span class="no">cdb</span>, <span class="kw">sequence_key</span> <span class="kw">=</span> <span class="st">'cdr3_nt'</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'DNA'</span>)
<span class="co">#&gt; Calculating intradistances on 1342 clusters.</span>
<span class="co">#&gt; Summarizing</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">cdb</span>$<span class="no">cluster_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">n_cluster</span><span class="kw">&gt;</span><span class="fl">1</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(<span class="no">key</span>, <span class="no">value</span>, -<span class="no">DNA97</span>) , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">value</span>))+ <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">key</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">'free'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We can also cluster by DNA identity.</p>
</div>
<div id="cluster-by-v-j-identity" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-by-v-j-identity" class="anchor"></a>Cluster by V-J identity</h1>
<div class="sourceCode"><html><body><pre class="r"><span class="no">germline_cluster</span> <span class="kw">=</span> <span class="fu"><a href="../reference/cluster_germline.html">cluster_germline</a></span>(<span class="no">cdb</span>, <span class="kw">segment_keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'chain'</span>), <span class="kw">cluster_name</span> <span class="kw">=</span> <span class="st">'segment_idx'</span>)
<span class="co">#&gt; Warning in replace_cluster_tbl(ccdb, cluster_tbl, cl_con_tbl, cluster_pk =</span>
<span class="co">#&gt; cluster_name): Replacing `cluster_tbl` with DNA97.</span></pre></body></html></div>
<p>We can cluster by any other feature of the contigs. Here we cluster each contig based on the chain and V-J genes. This gives us the set of observed V-J pairings:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">germline_cluster</span> <span class="kw">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(<span class="no">germline_cluster</span>, <span class="kw">sequence_key</span> <span class="kw">=</span> <span class="st">'cdr3_nt'</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'DNA'</span>)
<span class="co">#&gt; Calculating intradistances on 700 clusters.</span>
<span class="co">#&gt; Summarizing</span>
<span class="co">#&gt; Warning in left_join_warn(d_medoid, contig_tbl, by = ccdb$contig_pk, overwrite =</span>
<span class="co">#&gt; TRUE): Overwriting fields d(medoid), is_medoid in table contig_tbl</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">germline_cluster</span>$<span class="no">cluster_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">chain</span> <span class="kw">==</span> <span class="st">'TRB'</span>), <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">v_gene</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">j_gene</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">n_cluster</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">90</span>))</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Number of pairs</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">germline_cluster</span>$<span class="no">cluster_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">chain</span> <span class="kw">==</span> <span class="st">'TRB'</span>), <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">v_gene</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">j_gene</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">avg_distance</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">90</span>))</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Average Levenshtein distance of CDR3 within each pair. This might be turned into a z-score by fitting a weighted linear model with sum-to-zero contrasts and returning the pearson residuals. This could determine if a pairing has an unexpected small, or large, within cluster distance.</p>
</div>
<div id="oligo-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#oligo-clusters" class="anchor"></a>Oligo clusters</h1>
<p>Next, we will examine the clusters that are found in many contigs. First we will get a canonical contig to represent each cluster. This will be the medoid contig, by default.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">aa80</span> <span class="kw">=</span> <span class="fu"><a href="../reference/canonicalize_cluster.html">canonicalize_cluster</a></span>(<span class="no">aa80</span>, <span class="kw">representative</span> <span class="kw">=</span> <span class="st">'cdr3'</span>, <span class="kw">contig_fields</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'cdr3_nt'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'d_gene'</span>, <span class="st">'j_gene'</span>))
<span class="co">#&gt; Filtering `contig_tbl` by `is_medoid`, override by setting `contig_filter_args == TRUE`</span></pre></body></html></div>
<p><code>aa80</code> now includes the fields listed in <code>contig_fields</code> in the <code>cluster_tbl</code>, using the values found in the medoid contig.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">MIN_OLIGO</span> <span class="kw">=</span> <span class="fl">7</span>
<span class="no">oligo_clusters</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">aa80</span>$<span class="no">cluster_tbl</span>, <span class="no">n_cluster</span> <span class="kw">&gt;=</span> <span class="no">MIN_OLIGO</span>)
<span class="no">oligo_contigs</span> <span class="kw">=</span> <span class="no">aa80</span>
<span class="no">oligo_contigs</span>$<span class="no">contig_tbl</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join.html">semi_join</a></span>(<span class="no">oligo_contigs</span>$<span class="no">contig_tbl</span>, <span class="no">oligo_clusters</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="st">'aa80'</span>)
<span class="no">oligo_contigs</span>
<span class="co">#&gt; ContigCellDB of 54 contigs; 832 cells; and 4 clusters.</span>
<span class="co">#&gt; Contigs keyed by barcode, pop, sample, contig_id; cells keyed by barcode, pop, sample.</span></pre></body></html></div>
<p>Get contigs/cells/clusters found at least 7 times (across contigs). Note that replacing <code>contig_tbl</code> with the subset selected with the <code>semi_join</code> also automatically subsetted the <code>cell_tbl</code> and <code>cluster_tbl</code>.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">oligo_clusters</span> <span class="kw">=</span> <span class="no">oligo_contigs</span>$<span class="no">contig_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">aa80</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="kw">`n subjects observed`</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">sample</span>))) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="no">oligo_clusters</span>)
<span class="co">#&gt; Joining, by = "aa80"</span>

<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="no">oligo_clusters</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">aa80</span>:<span class="no">cdr3</span>, <span class="no">chain</span>:<span class="no">j_gene</span>, <span class="no">avg_distance</span>, <span class="no">n_cluster</span>))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="right">aa80</th>
<th align="right">n subjects observed</th>
<th align="left">cdr3</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="right">avg_distance</th>
<th align="right">n_cluster</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">111</td>
<td align="right">6</td>
<td align="left">CVVGDRGSALGRLHF</td>
<td align="left">TRA</td>
<td align="left">TRAV11</td>
<td align="left">None</td>
<td align="left">TRAJ18</td>
<td align="right">0.6071429</td>
<td align="right">28</td>
</tr>
<tr class="even">
<td align="right">172</td>
<td align="right">5</td>
<td align="left">CAVSRASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV9N-3</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">2.1111111</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="right">296</td>
<td align="right">6</td>
<td align="left">CAASASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV14D-2</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">1.5000000</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="right">808</td>
<td align="right">4</td>
<td align="left">CATGNYAQGLTF</td>
<td align="left">TRA</td>
<td align="left">TRAV8D-2</td>
<td align="left">None</td>
<td align="left">TRAJ26</td>
<td align="right">1.3333333</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p>Report some statistics about these expanded clusters.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">oligo_plot</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">oligo_contigs</span>$<span class="no">contig_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">representative</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">chain</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="kw">type</span> <span class="kw">=</span> <span class="st">'qual'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()
<span class="no">oligo_plot</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>These always come from a single chain.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">oligo_plot</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">fill</span> <span class="kw">=</span>   <span class="no">sample</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">pop</span>)</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>But come from multiple populations and samples.</p>
<div id="some-simple-phylogenetic-relationships" class="section level2">
<h2 class="hasAnchor">
<a href="#some-simple-phylogenetic-relationships" class="anchor"></a>Some simple phylogenetic relationships</h2>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggdendro</span>)

<span class="co"># This should be turned into a function in the package somehow</span>
<span class="co"># But plot arguments will be super-variable</span>
<span class="co"># Maybe just return the `hc` object?</span>
<span class="no">dendro_plot</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">ccdb</span>, <span class="no">idx</span>, <span class="no">method</span> <span class="kw">=</span> <span class="st">'complete'</span>){
    <span class="no">h</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">ccdb</span>$<span class="no">cluster_tbl</span>, !!<span class="fu"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(<span class="no">ccdb</span>$<span class="no">cluster_pk</span>) <span class="kw">==</span> <span class="no">idx</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span>(<span class="no">fc</span>) <span class="kw">%&gt;%</span> <span class="no">.</span><span class="kw">[[</span><span class="fl">1</span>]]
    <span class="no">quer</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">ccdb</span>$<span class="no">contig_tbl</span>, !!<span class="fu"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(<span class="no">ccdb</span>$<span class="no">cluster_pk</span>) <span class="kw">==</span> <span class="no">idx</span>)
    <span class="no">hc</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="no">h</span>$<span class="no">distance_mat</span>), <span class="kw">method</span> <span class="kw">=</span> <span class="no">method</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/dendro_data.html">dendro_data</a></span>(<span class="kw">type</span> <span class="kw">=</span> <span class="st">"rectangle"</span>)
    <span class="no">hc</span>$<span class="no">labels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">hc</span>$<span class="no">labels</span>, <span class="no">quer</span>)
   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">hc</span>$<span class="no">segments</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">x</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">y</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">xend</span><span class="kw">=</span><span class="no">xend</span>, <span class="kw">yend</span><span class="kw">=</span><span class="no">yend</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">hc</span>$<span class="no">labels</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">sample</span>, <span class="kw">label</span> <span class="kw">=</span> <span class="no">fancy_name</span>), <span class="kw">size</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">angle</span> <span class="kw">=</span> <span class="fl">60</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="kw">breaks</span> <span class="kw">=</span> <span class="kw">NULL</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'AA Distance'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">''</span>)
}

<span class="no">to_plot</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">aa80</span>$<span class="no">cluster_tbl</span>, <span class="no">n_cluster</span> <span class="kw">&gt;=</span> <span class="no">MIN_OLIGO</span>)

<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="no">to_plot</span>$<span class="no">aa80</span>, ~ <span class="fu">dendro_plot</span>(<span class="no">aa80</span>, <span class="no">.</span>))
<span class="co">#&gt; [[1]]</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-2.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-3.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-4.png" width="700"></p>
</div>
<div id="formal-testing-for-frequency-differences" class="section level2">
<h2 class="hasAnchor">
<a href="#formal-testing-for-frequency-differences" class="anchor"></a>Formal testing for frequency differences</h2>
<div class="sourceCode"><html><body><pre class="r"><span class="no">mm_out</span> <span class="kw">=</span> <span class="fu"><a href="../reference/cluster_logistic_test.html">cluster_test_by</a></span>(<span class="no">aa80</span>, <span class="kw">fields</span> <span class="kw">=</span> <span class="st">'chain'</span>, <span class="st">'cluster_tbl'</span>, ~ <span class="no">pop</span> + (<span class="fl">1</span><span class="kw">|</span><span class="no">sample</span>), <span class="kw">filterset</span> <span class="kw">=</span> <span class="fu"><a href="../reference/cluster_filterset.html">cluster_filterset</a></span>(<span class="kw">white_list</span> <span class="kw">=</span> <span class="no">oligo_clusters</span>)) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="no">oligo_clusters</span>)
<span class="co">#&gt; Fitting mixed logistic models to 4 clusters.</span>
<span class="co">#&gt; Loading required namespace: broom</span>
<span class="co">#&gt; Loading required namespace: lme4</span>
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; Fitting mixed logistic models to 0 clusters.</span>
<span class="co">#&gt; Joining, by = c("chain", "aa80")</span>

<span class="no">mm_out</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="no">mm_out</span>, <span class="kw">conf.low</span> <span class="kw">=</span> <span class="no">estimate</span>-<span class="fl">1.96</span>*<span class="no">std.error</span>, <span class="kw">conf.high</span> <span class="kw">=</span> <span class="no">estimate</span> + <span class="fl">1.96</span>*<span class="no">std.error</span>)</pre></body></html></div>
<div class="sourceCode"><html><body><pre class="r"><span class="no">mm_outj</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>(<span class="no">mm_out</span>), <span class="no">term</span> <span class="kw">==</span> <span class="st">'popbalbc'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(<span class="no">representative</span>))

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">mm_outj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">representative</span>, <span class="kw">ymin</span> <span class="kw">=</span> <span class="no">conf.low</span>, <span class="kw">ymax</span> <span class="kw">=</span> <span class="no">conf.high</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">estimate</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>()  + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="kw">yintercept</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">lty</span> <span class="kw">=</span> <span class="fl">2</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Isomorph"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"log odds of isomorph"</span>)</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/per_iso_tests-1.png" width="700"></p>
<p>We test if the binomial rate of clone expression differs between CD31+/- or term, for each clone.</p>
</div>
</div>
<div id="clonal-pairs" class="section level1">
<h1 class="hasAnchor">
<a href="#clonal-pairs" class="anchor"></a>Clonal pairs</h1>
<p>Next, we can examine the pairing between <span class="math inline">\(\alpha-\beta\)</span> chains and see if any pairs are found more than once.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">aa80</span>$<span class="no">cluster_pk</span> <span class="kw">=</span> <span class="st">'representative'</span>
<span class="no">pairing_list</span> <span class="kw">=</span> <span class="fu"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(<span class="no">aa80</span>, <span class="kw">table_order</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">orphan_level</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">min_expansion</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">cluster_keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))
<span class="co">#&gt; Warning: Factor `cluster_idx.2` contains implicit NA, consider using</span>
<span class="co">#&gt; `forcats::fct_explicit_na`</span>
<span class="co">#&gt; Warning: Column `representative` joining factors with different levels, coercing</span>
<span class="co">#&gt; to character vector</span></pre></body></html></div>
<p><code>pairing_tables</code> finds all contig combinations of order <code>table_order</code> across cells. Among those combinations that occur at least <code>min_expansion</code> times, the expanded combinations and and any other combinations that shared an expanded combo.</p>
<div class="sourceCode"><html><body><pre class="r">
<span class="no">pairs_plt</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">pairing_list</span>$<span class="no">cell_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">cluster_idx.1_fct</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">cluster_idx.2_fct</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">sample</span>, <span class="kw">shape</span> <span class="kw">=</span> <span class="no">pop</span>), <span class="kw">width</span> <span class="kw">=</span> <span class="fl">.2</span>, <span class="kw">height</span> <span class="kw">=</span> <span class="fl">.2</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">45</span>))

<span class="no">pairs_plt</span> <span class="kw">=</span> <span class="fu"><a href="../reference/map_axis_labels.html">map_axis_labels</a></span>(<span class="no">pairs_plt</span>, <span class="no">pairing_list</span>$<span class="no">idx1_tbl</span>, <span class="no">pairing_list</span>$<span class="no">idx2_tbl</span>, <span class="kw">aes_label</span>  <span class="kw">=</span> <span class="st">'chain'</span>)
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="no">pairs_plt</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/plot_expanded-1.png" width="700"></p>
<div id="expanded-clones" class="section level2">
<h2 class="hasAnchor">
<a href="#expanded-clones" class="anchor"></a>Expanded clones</h2>
<div class="sourceCode"><html><body><pre class="r"><span class="no">whitelist</span> <span class="kw">=</span> <span class="no">oligo_clusters</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">cluster_idx.1</span> <span class="kw">=</span> <span class="no">representative</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>()
<span class="no">pairing_list</span> <span class="kw">=</span> <span class="fu"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(<span class="no">aa80</span>, <span class="kw">canonicalize_fun</span> <span class="kw">=</span> <span class="no">canonicalize_by_prevalence</span>, <span class="kw">table_order</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">orphan_level</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">min_expansion</span> <span class="kw">=</span> <span class="fl">Inf</span>, <span class="kw">cluster_whitelist</span> <span class="kw">=</span> <span class="no">whitelist</span>,  <span class="kw">cluster_keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))
<span class="co">#&gt; Warning: Factor `cluster_idx.2` contains implicit NA, consider using</span>
<span class="co">#&gt; `forcats::fct_explicit_na`</span>
<span class="co">#&gt; Warning: Column `representative` joining factors with different levels, coercing</span>
<span class="co">#&gt; to character vector</span>

<span class="no">pairs_plt</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">pairing_list</span>$<span class="no">cell_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">cluster_idx.1_fct</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">cluster_idx.2_fct</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">sample</span>, <span class="kw">shape</span> <span class="kw">=</span> <span class="no">pop</span>), <span class="kw">width</span> <span class="kw">=</span> <span class="fl">.2</span>, <span class="kw">height</span> <span class="kw">=</span> <span class="fl">.2</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">45</span>))

<span class="no">pairs_plt</span> <span class="kw">=</span> <span class="fu"><a href="../reference/map_axis_labels.html">map_axis_labels</a></span>(<span class="no">pairs_plt</span>, <span class="no">pairing_list</span>$<span class="no">idx1_tbl</span>, <span class="no">pairing_list</span>$<span class="no">idx2_tbl</span>, <span class="kw">aes_label</span>  <span class="kw">=</span> <span class="st">'chain'</span>)
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="no">pairs_plt</span></pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>By setting <code>min_expansion = Inf, cluster_whitelist = whitelist</code> we can examine any pairings for a set of cluster_idx, in this case the ones that were seen multiple times. Interestingly (and unlike some human samples) the expanded clusters are <span class="math inline">\(\beta\)</span>-chain, and their <span class="math inline">\(\alpha\)</span> chains are sprinkled quite evenly across clusters.</p>
</div>
</div>
<div id="length-of-cdr3" class="section level1">
<h1 class="hasAnchor">
<a href="#length-of-cdr3" class="anchor"></a>Length of CDR3</h1>
<div class="sourceCode"><html><body><pre class="r"><span class="no">aa80</span>$<span class="no">contig_tbl</span> <span class="kw">=</span> <span class="no">aa80</span>$<span class="no">contig_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">cdr3_length</span> <span class="kw">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(<span class="no">cdr3_nt</span>))
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">aa80</span>$<span class="no">contig_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="no">pop</span>, <span class="kw">x</span><span class="kw">=</span> <span class="no">cdr3_length</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="kw">binwidth</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">mapping</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">y</span> <span class="kw">=</span> <span class="no">..density..</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="kw">type</span> <span class="kw">=</span> <span class="st">'qual'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(<span class="no">sample</span> ~<span class="no">chain</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">strip.text.y</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span> <span class="kw">=</span> <span class="fl">0</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">25</span>, <span class="fl">55</span>))</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Plot the CDR3 length distribution for each sample and pop. There doesn’t appear to be a noticeable difference between BALB/c and C57BL/6J (b6) mice, but if we needed to make sure, an appropriate procedure would be to run a mixed model with a random <code>sample</code> effect (assumed to represent a biological replicate).</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">cdr_len</span> <span class="kw">=</span> <span class="no">aa80</span>$<span class="no">contig_tbl</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">chain</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span>(<span class="kw pkg">broom</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(<span class="kw pkg">lme4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(<span class="no">cdr3_length</span> ~ <span class="no">pop</span> + (<span class="fl">1</span><span class="kw">|</span><span class="no">sample</span>), <span class="kw">data</span> <span class="kw">=</span> <span class="no">.</span>), <span class="kw">conf.int</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">cdr_len</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">group</span> <span class="kw">==</span> <span class="st">'fixed'</span>, <span class="no">term</span> <span class="kw">!=</span> <span class="st">'(Intercept)'</span>), <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span>(<span class="no">chain</span>, <span class="no">term</span>), <span class="kw">y</span> <span class="kw">=</span> <span class="no">estimate</span>, <span class="kw">ymin</span> <span class="kw">=</span> <span class="no">conf.low</span>, <span class="kw">ymax</span> <span class="kw">=</span> <span class="no">conf.high</span>)) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'Length(CDR3 Nt)'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'Term/Chain'</span>)</pre></body></html></div>
<p><img src="cdr3_clustering_files/figure-html/cdr3_len-1.png" width="288"></p>
<p>We end up with a (harmless) convergence warning about a singular fit. This is expected, because the <code>samples</code> aren’t actually replicates – they are just subsamples drawn for illustrative purposes. The Balbc mice have .5 fewer nucleotides per contig, on average, and this is not significant.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Yu Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
