<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering repertoire via CDR3 sequences • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering repertoire via CDR3 sequences">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.98.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering repertoire via CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Clustering repertoire via CDR3 sequences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/master/vignettes/cdr3_clustering.Rmd"><code>vignettes/cdr3_clustering.Rmd</code></a></small>
      <div class="hidden name"><code>cdr3_clustering.Rmd</code></div>

    </div>

    
    
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(CellaRepertorium)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(readr)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyr)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(stringr)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(purrr)</code></pre></div>
<div id="load-filtered-contig-files" class="section level1">
<h1 class="hasAnchor">
<a href="#load-filtered-contig-files" class="anchor"></a>Load filtered contig files</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(contigs_qc)
MIN_CDR3_AA =<span class="st"> </span><span class="dv">6</span>


cdb =<span class="st"> </span><span class="kw"><a href="../reference/ContigCellDB-fun.html">ContigCellDB_10XVDJ</a></span>(contigs_qc, <span class="dt">contig_pk =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>, <span class="st">'contig_id'</span>), <span class="dt">cell_pk =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>))

cdb$contig_tbl =<span class="st"> </span>dplyr::<span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(cdb$contig_tbl, full_length, productive ==<span class="st"> 'True'</span>, high_confidence, chain !=<span class="st"> 'Multi'</span>, <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(cdr3) &gt;<span class="st"> </span>MIN_CDR3_AA) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>( <span class="dt">fancy_name =</span> <span class="kw"><a href="../reference/fancy_name_contigs.html">fancy_name_contigs</a></span>(., <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_c.html">str_c</a></span>(pop, <span class="st">'_'</span>, sample)))</code></pre></div>
<p>832 good chains (either TRA or TRB); each cell can appear more than once.</p>
</div>
<div id="chain-pairings" class="section level1">
<h1 class="hasAnchor">
<a href="#chain-pairings" class="anchor"></a>Chain pairings</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">paired_chain =<span class="st"> </span><span class="kw"><a href="../reference/enumerate_pairing.html">enumerate_pairing</a></span>(cdb, <span class="dt">chain_recode_fun =</span> <span class="st">'guess'</span>)

<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(paired_chain, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span>(sample, pop), <span class="dt">fill =</span> pairing)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~canonical, <span class="dt">scale =</span> <span class="st">'free_x'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div id="cluster-cdr3-protein-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-protein-sequences" class="anchor"></a>Cluster CDR3 protein sequences</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
aa80 =<span class="st"> </span><span class="kw"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(cdb, <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">cluster_name =</span> <span class="st">'aa80'</span>, <span class="dt">identity =</span> .<span class="dv">8</span>)
aa80 =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(aa80, <span class="dt">sequence_key =</span> <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">keep_clustering_details =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; Calculating intradistances on 988 clusters.</span>
<span class="co">#&gt; Summarizing</span>


<span class="co"># This maybe should be a turned into a function </span>
<span class="co"># Other plots should be considered:</span>
<span class="co"># That show how clusters are split between samples, chains, etc</span>
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(aa80$cluster_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(n_cluster&gt;<span class="dv">1</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(key, value, -aa80, -fc) , <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> value))+<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~key, <span class="dt">scales =</span> <span class="st">'free'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We cluster the CDR3 translated amino acid residues with the program <a href="http://weizhongli-lab.org/cdhit_suite/cgi-bin/index.cgi?cmd=cd-hit">CD-HIT</a>. A sequence is included in a cluster if it matches by 100% similarity and has the same CDR3 length. Note that this can and should be relaxed – especially in the beta chain we see “near clones” that only differ by a residue or two, seemingly in stylized places.</p>
</div>
<div id="cluster-cdr3-dna-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-cdr3-dna-sequences" class="anchor"></a>Cluster CDR3 DNA sequences</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cdb =<span class="st"> </span><span class="kw"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(cdb, <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>, <span class="dt">cluster_name =</span> <span class="st">'DNA97'</span>, <span class="dt">identity =</span> .<span class="dv">965</span>, <span class="dt">min_length =</span> MIN_CDR3_AA*<span class="dv">3-1</span>, <span class="dt">G =</span> <span class="dv">1</span>)
cdb =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(cdb, <span class="dt">sequence_key =</span> <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>)
<span class="co">#&gt; Calculating intradistances on 1342 clusters.</span>
<span class="co">#&gt; Summarizing</span>
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(cdb$cluster_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(n_cluster&gt;<span class="dv">1</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(key, value, -DNA97) , <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> value))+<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~key, <span class="dt">scales =</span> <span class="st">'free'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_sqrt</a></span>()
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We can also cluster by DNA identity.</p>
</div>
<div id="cluster-by-v-j-identity" class="section level1">
<h1 class="hasAnchor">
<a href="#cluster-by-v-j-identity" class="anchor"></a>Cluster by V-J identity</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">germline_cluster =<span class="st"> </span><span class="kw"><a href="../reference/cluster_germline.html">cluster_germline</a></span>(cdb, <span class="dt">segment_keys =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'chain'</span>), <span class="dt">cluster_name =</span> <span class="st">'segment_idx'</span>)
<span class="co">#&gt; Warning in replace_cluster_tbl(ccdb, cluster_tbl, cl_con_tbl, cluster_pk =</span>
<span class="co">#&gt; cluster_name): Replacing `cluster_tbl` with DNA97.</span></code></pre></div>
<p>We can cluster by any other feature of the contigs. Here we cluster each contig based on the chain and V-J genes. This gives us the set of observed V-J pairings:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">germline_cluster =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(germline_cluster, <span class="dt">sequence_key =</span> <span class="st">'cdr3_nt'</span>, <span class="dt">type =</span> <span class="st">'DNA'</span>)
<span class="co">#&gt; Calculating intradistances on 700 clusters.</span>
<span class="co">#&gt; Summarizing</span>
<span class="co">#&gt; Warning in left_join_warn(d_medoid, contig_tbl, by = ccdb$contig_pk, overwrite =</span>
<span class="co">#&gt; TRUE): Overwriting fields d(medoid), is_medoid in table contig_tbl</span>
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(germline_cluster$cluster_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(chain ==<span class="st"> 'TRB'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> v_gene, <span class="dt">y =</span> j_gene, <span class="dt">fill =</span> n_cluster)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>))</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Number of pairs</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(germline_cluster$cluster_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(chain ==<span class="st"> 'TRB'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> v_gene, <span class="dt">y =</span> j_gene, <span class="dt">fill =</span> avg_distance)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>))</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Average Levenshtein distance of CDR3 within each pair. This might be turned into a z-score by fitting a weighted linear model with sum-to-zero contrasts and returning the pearson residuals. This could determine if a pairing has an unexpected small, or large, within cluster distance.</p>
</div>
<div id="oligo-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#oligo-clusters" class="anchor"></a>Oligo clusters</h1>
<p>Next, we will examine the clusters that are found in many contigs. First we will get a canonical contig to represent each cluster. This will be the medoid contig, by default.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aa80 =<span class="st"> </span><span class="kw"><a href="../reference/canonicalize_cluster.html">canonicalize_cluster</a></span>(aa80, <span class="dt">representative =</span> <span class="st">'cdr3'</span>, <span class="dt">contig_fields =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'cdr3_nt'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'d_gene'</span>, <span class="st">'j_gene'</span>))
<span class="co">#&gt; Filtering `contig_tbl` by `is_medoid`, override by setting `contig_filter_args == TRUE`</span></code></pre></div>
<p><code>aa80</code> now includes the fields listed in <code>contig_fields</code> in the <code>cluster_tbl</code>, using the values found in the medoid contig.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MIN_OLIGO =<span class="st"> </span><span class="dv">7</span>
oligo_clusters =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(aa80$cluster_tbl, n_cluster &gt;=<span class="st"> </span>MIN_OLIGO)
oligo_contigs =<span class="st"> </span>aa80
oligo_contigs$contig_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">semi_join</a></span>(oligo_contigs$contig_tbl, oligo_clusters, <span class="dt">by =</span> <span class="st">'aa80'</span>)
oligo_contigs
<span class="co">#&gt; ContigCellDB of 54 contigs; 832 cells; and 4 clusters.</span>
<span class="co">#&gt; Contigs keyed by barcode, pop, sample, contig_id; cells keyed by barcode, pop, sample.</span></code></pre></div>
<p>Get contigs/cells/clusters found at least 7 times (across contigs). Note that replacing <code>contig_tbl</code> with the subset selected with the <code>semi_join</code> also automatically subsetted the <code>cell_tbl</code> and <code>cluster_tbl</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">oligo_clusters =<span class="st"> </span>oligo_contigs$contig_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(aa80) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="st">`</span><span class="dt">n subjects observed</span><span class="st">`</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(sample))) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(oligo_clusters)
<span class="co">#&gt; Joining, by = "aa80"</span>

knitr::<span class="kw">kable</span>(oligo_clusters %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(aa80:cdr3, chain:j_gene, avg_distance, n_cluster))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">aa80</th>
<th align="right">n subjects observed</th>
<th align="left">cdr3</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">d_gene</th>
<th align="left">j_gene</th>
<th align="right">avg_distance</th>
<th align="right">n_cluster</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">111</td>
<td align="right">6</td>
<td align="left">CVVGDRGSALGRLHF</td>
<td align="left">TRA</td>
<td align="left">TRAV11</td>
<td align="left">None</td>
<td align="left">TRAJ18</td>
<td align="right">0.6071429</td>
<td align="right">28</td>
</tr>
<tr class="even">
<td align="right">172</td>
<td align="right">5</td>
<td align="left">CAVSRASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV9N-3</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">2.1111111</td>
<td align="right">9</td>
</tr>
<tr class="odd">
<td align="right">296</td>
<td align="right">6</td>
<td align="left">CAASASSGSWQLIF</td>
<td align="left">TRA</td>
<td align="left">TRAV14D-2</td>
<td align="left">None</td>
<td align="left">TRAJ22</td>
<td align="right">1.5000000</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="right">808</td>
<td align="right">4</td>
<td align="left">CATGNYAQGLTF</td>
<td align="left">TRA</td>
<td align="left">TRAV8D-2</td>
<td align="left">None</td>
<td align="left">TRAJ26</td>
<td align="right">1.3333333</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p>Report some statistics about these expanded clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">oligo_plot =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(oligo_contigs$contig_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> representative, <span class="dt">fill =</span> chain)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">type =</span> <span class="st">'qual'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()
oligo_plot</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>These always come from a single chain.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">oligo_plot +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">fill =</span>   sample) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~pop)</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>But come from multiple populations and samples.</p>
<div id="some-simple-phylogenetic-relationships" class="section level2">
<h2 class="hasAnchor">
<a href="#some-simple-phylogenetic-relationships" class="anchor"></a>Some simple phylogenetic relationships</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggdendro)

<span class="co"># This should be turned into a function in the package somehow</span>
<span class="co"># But plot arguments will be super-variable</span>
<span class="co"># Maybe just return the `hc` object?</span>
dendro_plot =<span class="st"> </span>function(ccdb, idx, <span class="dt">method =</span> <span class="st">'complete'</span>){
    h =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(ccdb$cluster_tbl, !!<span class="kw"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(ccdb$cluster_pk) ==<span class="st"> </span>idx) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span>(fc) %&gt;%<span class="st"> </span>.[[<span class="dv">1</span>]]
    quer =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(ccdb$contig_tbl, !!<span class="kw"><a href="https://dplyr.tidyverse.org/reference/tidyeval.html">sym</a></span>(ccdb$cluster_pk) ==<span class="st"> </span>idx)
    hc =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(h$distance_mat), <span class="dt">method =</span> method) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ggdendro/man/dendro_data.html">dendro_data</a></span>(<span class="dt">type =</span> <span class="st">"rectangle"</span>)
    hc$labels =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(hc$labels, quer)
   <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(hc$segments, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>x, <span class="dt">y=</span>y)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">xend=</span>xend, <span class="dt">yend=</span>yend)) +<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span>(<span class="dt">data =</span> hc$labels, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">color =</span> sample, <span class="dt">label =</span> fancy_name), <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">angle =</span> <span class="dv">60</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">breaks =</span> <span class="ot">NULL</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'AA Distance'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">''</span>)
}

to_plot =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(aa80$cluster_tbl, n_cluster &gt;=<span class="st"> </span>MIN_OLIGO)

<span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(to_plot$aa80, ~<span class="st"> </span><span class="kw">dendro_plot</span>(aa80, .))
<span class="co">#&gt; [[1]]</span></code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-2.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-3.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-14-4.png" width="700"></p>
</div>
<div id="formal-testing-for-frequency-differences" class="section level2">
<h2 class="hasAnchor">
<a href="#formal-testing-for-frequency-differences" class="anchor"></a>Formal testing for frequency differences</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(lme4)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(broom)
per_chain_sample =<span class="st"> </span>good_cluster_cells %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(sample, pop, chain) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="dt">total_cells =</span> <span class="kw"><a href="https://dplyr.tidyverse.org/reference/n.html">n</a></span>(), <span class="dt">weeks_premature =</span> weeks_premature[<span class="dv">1</span>])

oligo_cluster_stat =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">semi_join</a></span>(oligo_clusters, good_cluster_cells %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(dataset, contig_id)) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(sample, pop, chain, cluster_idx) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="dt">n_cluster =</span> <span class="kw"><a href="https://dplyr.tidyverse.org/reference/n.html">n</a></span>())%&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>() %&gt;%<span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/complete.html">complete</a></span>(sample, pop, <span class="kw"><a href="https://tidyr.tidyverse.org/reference/expand.html">nesting</a></span>(cluster_idx, chain), <span class="dt">fill =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">n_cluster =</span> <span class="dv">0</span>))

oligo_cluster_stat =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(oligo_cluster_stat, per_chain_sample, <span class="dt">by =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'sample'</span>, <span class="st">'pop'</span>, <span class="st">'chain'</span>)) 

<span class="kw"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span>( <span class="kw"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(oligo_cluster_stat, <span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(chain, cluster_idx)) &gt;<span class="st"> </span><span class="dv">0</span>) ==<span class="st"> </span><span class="dv">1</span>))

mm_out =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span>(oligo_cluster_stat %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(cluster_idx, chain) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span>( <span class="kw"><a href="https://rdrr.io/pkg/lme4/man/glmer.html">glmer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(n_cluster, total_cells) ~<span class="st"> </span>pop +<span class="st"> </span>weeks_premature +<span class="st"> </span>(<span class="dv">1</span>|sample), <span class="dt">data =</span> ., <span class="dt">family =</span> <span class="st">'binomial'</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(<span class="dt">conf.int =</span> <span class="ot">TRUE</span>)))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mm_outj =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span>(mm_out), <span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(oligo_clusters_all %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(representative, cluster_idx))), term %in%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'popCD31Pos'</span>, <span class="st">'weeks_premature'</span>)) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">ci_lo =</span> AMmisc::<span class="kw">clamp</span>(conf.low), <span class="dt">ci_hi =</span> AMmisc::<span class="kw">clamp</span>(conf.high)) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(representative))

<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(mm_outj, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> representative, <span class="dt">ymin =</span> ci_lo, <span class="dt">ymax =</span> ci_hi, <span class="dt">y =</span> <span class="kw">clamp</span>(estimate))) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~term, <span class="dt">scales =</span> <span class="st">'free'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">lty =</span> <span class="dv">2</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Isomorph"</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"log odds of isomorph"</span>)</code></pre></div>
<p>We test if the binomial rate of clone expression differs between CD31+/- or term, for each clone.</p>
</div>
</div>
<div id="clonal-pairs" class="section level1">
<h1 class="hasAnchor">
<a href="#clonal-pairs" class="anchor"></a>Clonal pairs</h1>
<p>Next, we can examine the pairing between <span class="math inline">\(\alpha-\beta\)</span> chains and see if any pairs are found more than once.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">class_colors =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">tibble</a></span>(<span class="dt">chain =</span>  <span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(aa80$cluster_tbl$chain)) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span>  RColorBrewer::<span class="kw">brewer.pal</span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(chain),<span class="st">"Set1"</span>)[<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(chain)])
<span class="co">#&gt; Warning in RColorBrewer::brewer.pal(length(chain), "Set1"): minimal value for n is 3, returning requested palette with 3 different levels</span>

aa80$cluster_pk =<span class="st"> 'representative'</span>
pairing_list =<span class="st"> </span><span class="kw"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(aa80, <span class="dt">table_order =</span> <span class="dv">2</span>, <span class="dt">orphan_level =</span> <span class="dv">1</span>, <span class="dt">min_expansion =</span> <span class="dv">3</span>, <span class="dt">cluster_keys =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))
<span class="co">#&gt; Warning: Factor `cluster_idx.2` contains implicit NA, consider using</span>
<span class="co">#&gt; `forcats::fct_explicit_na`</span>
<span class="co">#&gt; Warning: Column `representative` joining factors with different levels, coercing</span>
<span class="co">#&gt; to character vector</span></code></pre></div>
<p><code>pairing_tables</code> finds all contig combinations of order <code>table_order</code> across cells. Among those combinations that occur at least <code>min_expansion</code> times, the expanded combinations and and any other combinations that shared an expanded combo.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pairs_plt =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pairing_list$cell_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> cluster_idx.1_fct, <span class="dt">y =</span> cluster_idx.2_fct, <span class="dt">color =</span> sample, <span class="dt">shape =</span> pop)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="dt">width =</span> .<span class="dv">2</span>, <span class="dt">height =</span> .<span class="dv">2</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>)

feature_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(pairing_list$idx1_tbl, pairing_list$idx2_tbl), class_colors)
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; Joining, by = "chain"</span>

ylab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)$layout$panel_params[[<span class="dv">1</span>]]$y.label) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))
<span class="co">#&gt; Warning: `data_frame()` is deprecated, use `tibble()`.</span>
<span class="co">#&gt; This warning is displayed once per session.</span>
<span class="co">#&gt; Joining, by = "representative"</span>

xlab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)$layout$panel_params[[<span class="dv">1</span>]]$x.label) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))
<span class="co">#&gt; Joining, by = "representative"</span>

pairs_plt =<span class="st"> </span>pairs_plt +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">color =</span> xlab$class_color, <span class="dt">size =</span> <span class="dv">8</span>), <span class="dt">axis.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">color =</span> ylab$class_color, <span class="dt">size =</span> <span class="dv">8</span>))

pairs_plt</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/plot_expanded-1.png" width="700"></p>
<div id="expanded-clones" class="section level2">
<h2 class="hasAnchor">
<a href="#expanded-clones" class="anchor"></a>Expanded clones</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">whitelist =<span class="st"> </span>oligo_clusters %&gt;%<span class="st"> </span>dplyr::<span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="dt">cluster_idx.1 =</span> representative) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>()
pairing_list =<span class="st"> </span><span class="kw"><a href="../reference/pairing_tables.html">pairing_tables</a></span>(aa80, <span class="dt">canonicalize_fun =</span> canonicalize_by_prevalence, <span class="dt">table_order =</span> <span class="dv">2</span>, <span class="dt">orphan_level =</span> <span class="dv">1</span>, <span class="dt">min_expansion =</span> <span class="ot">Inf</span>, <span class="dt">cluster_whitelist =</span> whitelist,  <span class="dt">cluster_keys =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span>))
pairs_plt =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pairing_list$cell_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> cluster_idx.1_fct, <span class="dt">y =</span> cluster_idx.2_fct, <span class="dt">color =</span> sample, <span class="dt">shape =</span> pop)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="dt">width =</span> .<span class="dv">2</span>, <span class="dt">height =</span> .<span class="dv">2</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'TRB'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'TRA'</span>)

feature_tbl =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(pairing_list$idx1_tbl, pairing_list$idx2_tbl), class_colors)

ylab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)$layout$panel_params[[<span class="dv">1</span>]]$y.label) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))

xlab =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">data_frame</a></span>(<span class="dt">representative =</span>  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot_build.html">ggplot_build</a></span>(pairs_plt)$layout$panel_params[[<span class="dv">1</span>]]$x.label) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(feature_tbl) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">class_color =</span> <span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(class_color), <span class="st">'#E41A1C'</span>, class_color))

pairs_plt =<span class="st"> </span>pairs_plt +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">color =</span> xlab$class_color, <span class="dt">size =</span> <span class="dv">8</span>), <span class="dt">axis.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">color =</span> ylab$class_color, <span class="dt">size =</span> <span class="dv">8</span>))

pairs_plt</code></pre></div>
<p>By setting <code>min_expansion = Inf, cluster_whitelist = whitelist</code> we can examine any pairings for a set of cluster_idx, in this case the ones that were seen multiple times. Interestingly (and unlike some human samples) the expanded clusters are <span class="math inline">\(\beta\)</span>-chain, and their <span class="math inline">\(\alpha\)</span> chains are sprinkled quite evenly across clusters.</p>
</div>
</div>
<div id="length-of-cdr3" class="section level1">
<h1 class="hasAnchor">
<a href="#length-of-cdr3" class="anchor"></a>Length of CDR3</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aa80$contig_tbl =<span class="st"> </span>aa80$contig_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">cdr3_length =</span> <span class="kw"><a href="https://stringr.tidyverse.org/reference/str_length.html">str_length</a></span>(cdr3_nt))
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(aa80$contig_tbl, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">fill =</span> pop, <span class="dt">x=</span> cdr3_length)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">binwidth =</span> <span class="dv">1</span>, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> ..density..)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">type =</span> <span class="st">'qual'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(sample ~chain) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">strip.text.y =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">0</span>)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">xlim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">25</span>, <span class="dv">55</span>))</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Plot the CDR3 length distribution for each sample and pop. There doesn’t appear to be a noticeable difference between BALB/c and C57BL/6J (b6) mice, but if we needed to make sure, an appropriate procedure would be to run a mixed model with a random <code>sample</code> effect (assumed to represent a biological replicate).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cdr_len =<span class="st"> </span>aa80$contig_tbl %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(chain) %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span>(broom::<span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(lme4::<span class="kw"><a href="https://rdrr.io/pkg/lme4/man/lmer.html">lmer</a></span>(cdr3_length ~<span class="st"> </span>pop +<span class="st"> </span>(<span class="dv">1</span>|sample), <span class="dt">data =</span> .), <span class="dt">conf.int =</span> <span class="ot">TRUE</span>))
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="co">#&gt; boundary (singular) fit: see ?isSingular</span>
<span class="co">#&gt; Warning in bind_rows_(x, .id): binding factor and character vector, coercing</span>
<span class="co">#&gt; into character vector</span>

<span class="co">#&gt; Warning in bind_rows_(x, .id): binding character and factor vector, coercing</span>
<span class="co">#&gt; into character vector</span>
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(cdr_len %&gt;%<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(group ==<span class="st"> 'fixed'</span>, term !=<span class="st"> '(Intercept)'</span>), <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span>(chain, term), <span class="dt">y =</span> estimate, <span class="dt">ymin =</span> conf.low, <span class="dt">ymax =</span> conf.high)) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'Length(CDR3 Nt)'</span>) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'Term/Chain'</span>)</code></pre></div>
<p><img src="cdr3_clustering_files/figure-html/cdr3_len-1.png" width="288"></p>
<p>We end up with a (harmless) convergence warning about a singular fit. This is expected, because the <code>samples</code> aren’t actually replicates – they are just subsamples drawn for illustrative purposes. The Balbc mice have .5 fewer nucleotides per contig, on average, and this is not significant.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-filtered-contig-files">Load filtered contig files</a></li>
      <li><a href="#chain-pairings">Chain pairings</a></li>
      <li><a href="#cluster-cdr3-protein-sequences">Cluster CDR3 protein sequences</a></li>
      <li><a href="#cluster-cdr3-dna-sequences">Cluster CDR3 DNA sequences</a></li>
      <li><a href="#cluster-by-v-j-identity">Cluster by V-J identity</a></li>
      <li>
<a href="#oligo-clusters">Oligo clusters</a><ul class="nav nav-pills nav-stacked">
<li><a href="#some-simple-phylogenetic-relationships">Some simple phylogenetic relationships</a></li>
      <li><a href="#formal-testing-for-frequency-differences">Formal testing for frequency differences</a></li>
      </ul>
</li>
      <li>
<a href="#clonal-pairs">Clonal pairs</a><ul class="nav nav-pills nav-stacked">
<li><a href="#expanded-clones">Expanded clones</a></li>
      </ul>
</li>
      <li><a href="#length-of-cdr3">Length of CDR3</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Yu Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
