<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An Introduction to CellaRepertorium • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An Introduction to CellaRepertorium">
<meta property="og:description" content="CellaRepertorium">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering and differential usage of repertoire CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/cr-overview.html">An Introduction to CellaRepertorium</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An Introduction to CellaRepertorium</h1>
                        <h4 data-toc-skip class="author">Andrew
McDavid</h4>
            <address class="author_afil">
      University of Rochester, Department of Biostatistics and
Computational
Biology<br><a class="author_email" href="mailto:#"></a><a href="mailto:Andrew_McDavid@urmc.rochester.edu" class="email">Andrew_McDavid@urmc.rochester.edu</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/HEAD/vignettes/cr-overview.Rmd" class="external-link"><code>vignettes/cr-overview.Rmd</code></a></small>
      <div class="hidden name"><code>cr-overview.Rmd</code></div>

    </div>

    
    
<p>Congratulations on your installation of CellaRepertorium. This
package contains methods for manipulating, clustering, pairing, testing
and conducting multimodal analysis single cell RepSeq data, especially
as generated by <a href="https://support.10xgenomics.com/single-cell-vdj/software/overview/welcome" class="external-link">10X
Genomics Chromium Immune Profiling</a>.</p>
<div class="section level2">
<h2 id="ethos-and-data-structure">Ethos and data structure<a class="anchor" aria-label="anchor" href="#ethos-and-data-structure"></a>
</h2>
<p>The fundamental unit this package operates on is the
<strong>contig</strong>, which is a section of contiguously stitched
reads from a <strong>single cell</strong>. Each contig belongs to one
(and only one) cell, however, cells may generate multiple contigs.</p>
<p><img src="../../../../../private/var/folders/7m/c_l3kv2d2z118lwkj5jpy2h40000gn/T/RtmpTNeai5/temp_libpathf7cb3d4472ec/CellaRepertorium/help/figures/contig_schematic.png" width="590"></p>
<p>Contigs belong to cells, and can also belong to a
<strong>cluster</strong>. A <code><a href="../reference/ContigCellDB-fun.html">ContigCellDB()</a></code> object tracks
these two types of membership by using a sequence of three
<code>data.frames</code> (<code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">dplyr::tibble()</a></code>, actually).
<code><a href="../reference/ContigCellDB-fun.html">ContigCellDB()</a></code> also tracks columns (the primary keys) that
uniquely identify each row in each of these tables. The
<code>contig_tbl</code> is the <code>tibble</code> containing
<strong>contigs</strong>, the <code>cell_tbl</code> contains the
<strong>cells</strong>, and the <code>cluster_tbl</code> contains the
<strong>clusters</strong>.</p>
<p><img src="../../../../../private/var/folders/7m/c_l3kv2d2z118lwkj5jpy2h40000gn/T/RtmpTNeai5/temp_libpathf7cb3d4472ec/CellaRepertorium/help/figures/table_schematic.png" width="1010"></p>
<p>The <code>contig_pk</code>, <code>cell_pk</code> and
<code>cluster_pk</code> specify the columns that identify a contig, cell
and cluster, respectively. These will serve as foreign keys that link
the three tables together.</p>
</div>
<div class="section level2">
<h2 id="manipulation">Manipulation<a class="anchor" aria-label="anchor" href="#manipulation"></a>
</h2>
<p>We’ll start <em>in media res</em> with an example of
minimially-processed annotated contigs from Cellranger. For details on
how to import your own CellRanger data, as well QC steps that could be
performed, see <code><a href="../articles/mouse_tcell_qc.html">vignette('mouse_tcell_qc')</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/amcdavid/CellaRepertorium" class="external-link">CellaRepertorium</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"contigs_qc"</span><span class="op">)</span>
<span class="va">cdb</span> <span class="op">=</span> <span class="fu"><a href="../reference/ContigCellDB-fun.html">ContigCellDB_10XVDJ</a></span><span class="op">(</span><span class="va">contigs_qc</span>, 
                          contig_pk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span>, <span class="st">'contig_id'</span><span class="op">)</span>, 
                          cell_pk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'barcode'</span>, <span class="st">'pop'</span>, <span class="st">'sample'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This constructs a <code>ContigCellDB</code> object, specifying that
the columns <code>barcode</code>, <code>pop</code>, <code>sample</code>,
and <code>contig_id</code> unique identify a contig, so are its
<strong>primary keys</strong>, and that columns <code>barcode</code>,
<code>pop</code>, <code>sample</code> are the cell primary keys.</p>
<p>We can manipulate the <code>contig_tbl</code> with the <code>$</code>
operator.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdb</span><span class="op">$</span><span class="va">contig_tbl</span><span class="op">$</span><span class="va">cdr_nt_len</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">$</span><span class="va">contig_tbl</span><span class="op">$</span><span class="va">cdr3_nt</span><span class="op">)</span></code></pre></div>
<p>Or with the <code>mutate_cdb</code> function, which saves a few
keystrokes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span><span class="op">)</span>
<span class="va">cdb</span> <span class="op">=</span> <span class="va">cdb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/mutate_cdb.html">mutate_cdb</a></span><span class="op">(</span>cdr3_g_content <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html" class="external-link">alphabetFrequency</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">cdr3_nt</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="st">'G'</span><span class="op">]</span>, tbl <span class="op">=</span> <span class="st">'contig_tbl'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">$</span><span class="va">contig_tbl</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">contig_id</span>, <span class="va">cdr3_nt</span>, <span class="va">cdr_nt_len</span>, <span class="va">cdr3_g_content</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 4</span></span>
<span class="co">#&gt;   contig_id                   cdr3_nt                  cdr_nt_len cdr3_g_content</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AAAGTAGTCGCGCCAA-1_contig_1 TGTGCCAGCAGTCCGACAGACTA…         36              8</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> AAAGTAGTCGCGCCAA-1_contig_2 TGTGCCTGGAGTCCCGGGGACAA…         39             12</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> AAAGTAGTCGCGCCAA-1_contig_4 TGTGCTATAGAGGCAGGCAATAC…         39              9</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> AACCATGCATTTGCCC-1_contig_3 TGTGCTGTGAGCGCATACCAGGG…         42             14</span></code></pre></div>
<p>Other functionality, some of which is depicted in
<code><a href="../articles/mouse_tcell_qc.html">vignette('mouse_tcell_qc')</a></code> and
<code><a href="../articles/cdr3_clustering.html">vignette('cdr3_clustering')</a></code> includes:</p>
<ul>
<li>Exploring pairing (classical alpha-beta or heavy-light, single
chain, or &gt;1 chain type) with <code>enumerate_pairing</code>.</li>
<li>Splitting by a factor with <code>split_cdb</code>.</li>
<li>Filtering with <code>filter_cdb</code>.</li>
<li>Canonicalization with <code>canonicalize_cell</code>. This chooses a
representative contig for each cell and copies various fields into the
<code>cell_tbl</code> so that the contig-cell relationship in these
fields is now one-to-one. This is useful for any analysis of contigs
that requires the cell as its base. Likewise
<code>canonicalize_cluster</code> choses a representative contig for
each cluster.</li>
</ul>
</div>
<div class="section level2">
<h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<p>We provide an R port of the fast biostring clustering algorithm <a href="http://weizhongli-lab.org/cdhit_suite/cgi-bin/index.cgi?cmd=cd-hit" class="external-link">CD-HIT</a>
<span class="citation">(Fu et al. 2012)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span><span class="op">(</span><span class="va">cdb</span>, <span class="st">'cdr3'</span>, type <span class="op">=</span> <span class="st">'AA'</span>, cluster_pk <span class="op">=</span> <span class="st">'aa80'</span>, 
                  identity <span class="op">=</span> <span class="fl">.8</span>, min_length <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/fine_clustering.html">fine_clustering</a></span><span class="op">(</span><span class="va">aa80</span>, sequence_key <span class="op">=</span> <span class="st">'cdr3'</span>, type <span class="op">=</span> <span class="st">'AA'</span><span class="op">)</span>
<span class="co">#&gt; Calculating intradistances on 997 clusters.</span>
<span class="co">#&gt; Summarizing</span></code></pre></div>
<p>This partitions sequences into sets with &gt;80% mutual similarity in
the amino acid sequence, adds some additional information about the
clustering, and returns it as a <code>ContigCellDB</code> object named
<code>aa80</code>. The primary key for the clusters is aa80.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">cluster_tbl</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">#&gt;    aa80 avg_distance n_cluster</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     1            0         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     2            0         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     3            0         2</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span>     4            0         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span>     5            0         1</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span>     6            0         1</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">$</span><span class="va">contig_tbl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">contig_id</span>, <span class="va">aa80</span>, <span class="va">is_medoid</span>, <span class="va">`d(medoid)`</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">#&gt;   contig_id                    aa80 is_medoid `d(medoid)`</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ATCTACTCAGTATGCT-1_contig_3     1 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ACTGTCCTCAATCACG-1_contig_3     2 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> CACCTTGTCCAATGGT-1_contig_2     3 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CACCTTGTCCAATGGT-1_contig_2     3 FALSE               0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> CGGACGTGTTCATGGT-1_contig_1     4 TRUE                0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> CTGCTGTTCCCTAATT-1_contig_4     5 TRUE                0</span></code></pre></div>
<p>The <code>cluster_tbl</code> lists the 997 80% identity groups found,
including the number of contigs in the cluster, and the average distance
between elements in the group. In the <code>contig_tbl</code>, there are
two columns specifying if the contig <code>is_medoid</code>, that is, is
the most representative element of the set and the distance to the
medoid element <code>d(medoid)</code>.</p>
<p>Other functionality to operate on clustering include:</p>
<ul>
<li>
<code><a href="../reference/cluster_germline.html">cluster_germline()</a></code> defines clusters using combinations
of factors in the <code>contig_tbl</code>, such as the V- and J-gene
identities.</li>
</ul>
</div>
<div class="section level2">
<h2 id="pairing">Pairing<a class="anchor" aria-label="anchor" href="#pairing"></a>
</h2>
<p>One of the main benefits of single cell repertoire sequencing is the
ability to recover both light and heavy chains, or alpha and beta chains
of B cells and T cells. Pairing is a property of the
<code>cell_tbl</code>. We provide a number of tools to analyze and
visualize the pairing.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="va">paired_chain</span> <span class="op">=</span> <span class="fu"><a href="../reference/enumerate_pairing.html">enumerate_pairing</a></span><span class="op">(</span><span class="va">cdb</span>, chain_recode_fun <span class="op">=</span> <span class="st">'guess'</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">paired_chain</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html" class="external-link">interaction</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">pop</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">pairing</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">canonical</span>, scale <span class="op">=</span> <span class="st">'free_x'</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="cr-overview_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>We first determine how often cells were paired, and how often
<em>non-canonical</em> multi-alpha or multi-beta cells are found.<br>
T cells with two alpha chains, and to a lessor extent, two beta chains
are an established phenomena <span class="citation">(Padovan et al.
1993, 1995; He et al. 2002)</span>. However, if the rate of these
so-called dual TCR cells is too high, then multiplets or excess ambient
RNA may be suspected. (B plasma cells seem to be particularly sticky,
and are laden with immunoglobulin RNA).</p>
<div class="section level3">
<h3 id="enumerating-clonotypically-expanded-pairs">Enumerating clonotypically expanded pairs<a class="anchor" aria-label="anchor" href="#enumerating-clonotypically-expanded-pairs"></a>
</h3>
<p>Besides this high-level description of the rate and characteristics
of the pairing, we want to discover pairs that occur repeatedly. For
that, we can use the <code>pairing_tables</code> function.</p>
<p>First we copy some info into the <code>cluster_tbl</code> from the
medoid contig:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/canonicalize_cluster.html">canonicalize_cluster</a></span><span class="op">(</span><span class="va">aa80</span>, representative <span class="op">=</span> <span class="st">'cdr3'</span>, 
                            contig_fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cdr3'</span>, <span class="st">'cdr3_nt'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'d_gene'</span>, <span class="st">'j_gene'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Filtering `contig_tbl` by `is_medoid`, override by setting `contig_filter_args == TRUE`</span>
<span class="va">aa80</span><span class="op">$</span><span class="va">cluster_pk</span> <span class="op">=</span> <span class="st">'representative'</span></code></pre></div>
<p>The <code>representative</code> just gives the clusters a more useful
unique name (the CDR3 animo acid sequence). The other information would
be helpful with visualizing and understanding the results of the
pairing.</p>
<p>Next we provide an ordering for each contig in a cell:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aa80</span> <span class="op">=</span> <span class="fu"><a href="../reference/rank_prevalence_ccdb.html">rank_chain_ccdb</a></span><span class="op">(</span><span class="va">aa80</span><span class="op">)</span></code></pre></div>
<p>In this case the contigs will be picked by the beta/alpha chain of
the cluster. Other options are possible, for instance the prevalence of
the cluster with <code>rank_prevalence_ccdb</code>, which would allow
detection to see expanded, dual-TCR pairings (alpha-alpha or
beta-beta).</p>
<p>Finally, we generate the pairings.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pairing_list</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairing_tables.html">pairing_tables</a></span><span class="op">(</span><span class="va">aa80</span>, table_order <span class="op">=</span> <span class="fl">2</span>, orphan_level <span class="op">=</span> <span class="fl">1</span>, min_expansion <span class="op">=</span> <span class="fl">3</span>, cluster_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cdr3'</span>, <span class="st">'representative'</span>, <span class="st">'chain'</span>, <span class="st">'v_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'avg_distance'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>By default, this is subset to a list suitable for plotting in a
heatmap-like format, so includes only expanded pairings. You can get all
pairings by setting <code>min_expansion = 1</code>, and force inclusion
or exclusion of particular clusters with <code>cluster_whitelist</code>
or <code>cluster_blacklist</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pairing_list</span><span class="op">$</span><span class="va">cell_tbl</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cluster_idx.1_fct</span>, y <span class="op">=</span> <span class="va">cluster_idx.2_fct</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">sample</span>, shape <span class="op">=</span> <span class="va">pop</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.2</span>, height <span class="op">=</span> <span class="fl">.2</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'TRB'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'TRA'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span><span class="op">)</span><span class="op">)</span>

<span class="va">pairs_plt</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_axis_labels.html">map_axis_labels</a></span><span class="op">(</span><span class="va">pairs_plt</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx1_tbl</span>, <span class="va">pairing_list</span><span class="op">$</span><span class="va">idx2_tbl</span>, aes_label  <span class="op">=</span> <span class="st">'chain'</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: cowplot</span>
<span class="va">pairs_plt</span></code></pre></div>
<p><img src="cr-overview_files/figure-html/plot_expanded-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="testing">Testing<a class="anchor" aria-label="anchor" href="#testing"></a>
</h2>
<p>If the experiment that generated the data was a designed experiment,
it might be of interest to test clusters for differential abundance. We
implement ordinary and mixed-effect logistic and binomial tests. See
<code>cluster_logistic_test</code> for details and
<code><a href="../articles/cdr3_clustering.html">vignette('cdr3_clustering')</a></code> for an example.</p>
<p>We also implement a permutation test which may be suitable for
testing for various phylogenetic properties of clonotypes, such as the
clonotype diversity or polarization. Cluster assignments are permutated
by cells, possibly conditioning on less granular covariates. See
<code>cluster_permute_test</code> and
<code><a href="../articles/cdr3_clustering.html">vignette('cdr3_clustering')</a></code>.</p>
</div>
<div class="section level2">
<h2 id="multimodal-analysis">Multimodal Analysis<a class="anchor" aria-label="anchor" href="#multimodal-analysis"></a>
</h2>
<p><code>ContigCellDB</code> objects can be included as a field on the
<code>colData</code> of a <code>SingleCellExperiment</code>. This
permits various multimodal analyses, while maintaining the
correspondence between the cells in the <code>SinglCellExperiment</code>
and the cells in the <code>ContigCellDB</code>. See
<code><a href="../articles/repertoire_and_expression.html">vignette('repertoire_and_expression')</a></code>.</p>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>Development of CellaRepertorium was funded in part by UL1 TR002001
(PI Bennet/Zand) pilot to Andrew McDavid.</p>
</div>
<div class="section level2">
<h2 id="colophone">Colophone<a class="anchor" aria-label="anchor" href="#colophone"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Monterey 12.3.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] ggplot2_3.3.5          Biostrings_2.62.0      GenomeInfoDb_1.30.1   </span>
<span class="co">#&gt;  [4] XVector_0.34.0         IRanges_2.28.0         S4Vectors_0.32.3      </span>
<span class="co">#&gt;  [7] BiocGenerics_0.40.0    dplyr_1.0.8            CellaRepertorium_1.7.1</span>
<span class="co">#&gt; [10] BiocStyle_2.22.0      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.8             lattice_0.20-45        tidyr_1.2.0           </span>
<span class="co">#&gt;  [4] png_0.1-7              assertthat_0.2.1       rprojroot_2.0.2       </span>
<span class="co">#&gt;  [7] digest_0.6.29          utf8_1.2.2             plyr_1.8.6            </span>
<span class="co">#&gt; [10] R6_2.5.1               evaluate_0.15          highr_0.9             </span>
<span class="co">#&gt; [13] pillar_1.7.0           zlibbioc_1.40.0        rlang_1.0.2           </span>
<span class="co">#&gt; [16] rstudioapi_0.13        jquerylib_0.1.4        Matrix_1.3-4          </span>
<span class="co">#&gt; [19] rmarkdown_2.13         pkgdown_2.0.2          labeling_0.4.2        </span>
<span class="co">#&gt; [22] textshaping_0.3.6      desc_1.4.1             stringr_1.4.0         </span>
<span class="co">#&gt; [25] munsell_0.5.0          RCurl_1.98-1.6         compiler_4.1.2        </span>
<span class="co">#&gt; [28] xfun_0.30              pkgconfig_2.0.3        systemfonts_1.0.4     </span>
<span class="co">#&gt; [31] htmltools_0.5.2        tidyselect_1.1.2       tibble_3.1.6          </span>
<span class="co">#&gt; [34] GenomeInfoDbData_1.2.7 bookdown_0.25          fansi_1.0.2           </span>
<span class="co">#&gt; [37] withr_2.5.0            crayon_1.5.0           bitops_1.0-7          </span>
<span class="co">#&gt; [40] grid_4.1.2             gtable_0.3.0           jsonlite_1.8.0        </span>
<span class="co">#&gt; [43] lifecycle_1.0.1        DBI_1.1.2              magrittr_2.0.2        </span>
<span class="co">#&gt; [46] scales_1.1.1           cli_3.2.0              stringi_1.7.6         </span>
<span class="co">#&gt; [49] cachem_1.0.6           reshape2_1.4.4         farver_2.1.0          </span>
<span class="co">#&gt; [52] fs_1.5.2               bslib_0.3.1            ellipsis_0.3.2        </span>
<span class="co">#&gt; [55] ragg_1.2.2             generics_0.1.2         vctrs_0.3.8           </span>
<span class="co">#&gt; [58] cowplot_1.1.1          forcats_0.5.1          tools_4.1.2           </span>
<span class="co">#&gt; [61] glue_1.6.2             purrr_0.3.4            fastmap_1.1.0         </span>
<span class="co">#&gt; [64] yaml_2.3.5             colorspace_2.0-3       BiocManager_1.30.16   </span>
<span class="co">#&gt; [67] memoise_2.0.1          knitr_1.37             sass_0.4.1</span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Fu2012" class="csl-entry">
Fu, Limin, Beifang Niu, Zhengwei Zhu, Sitao Wu, and Weizhong Li. 2012.
<span>“<span class="nocase">CD-HIT: accelerated for clustering the
next-generation sequencing data.</span>”</span> <em>Bioinformatics
(Oxford, England)</em> 28 (23): 3150–52. <a href="https://doi.org/10.1093/bioinformatics/bts565" class="external-link">https://doi.org/10.1093/bioinformatics/bts565</a>.
</div>
<div id="ref-He2002" class="csl-entry">
He, Xin, Charles A. Janeway, Matthew Levine, Eve Robinson, Paula
Preston-Hurlburt, Christophe Viret, and Kim Bottomly. 2002. <span>“<span class="nocase">Dual receptor T cells extend the immune repertoire for
foreign antigens</span>.”</span> <em>Nature Immunology</em>. <a href="https://doi.org/10.1038/ni751" class="external-link">https://doi.org/10.1038/ni751</a>.
</div>
<div id="ref-Padovan1993" class="csl-entry">
Padovan, Elisabetta, Giulia Casorati, Paolo Dellabona, Stefan Meyer,
Manfred Brockhaus, and Antonio Lanzavecchia. 1993. <span>“<span class="nocase">Expression of two T cell receptor <span class="math inline">\(\alpha\)</span> chains: Dual receptor T
cells</span>.”</span> <em>Science</em>. <a href="https://doi.org/10.1126/science.8211163" class="external-link">https://doi.org/10.1126/science.8211163</a>.
</div>
<div id="ref-Padovan1995" class="csl-entry">
Padovan, Elisabetta, Claudia Giachino, Marina Cella, Salvatore
Valitutti, Oreste Acuto, and Antonio Lanzavecchia. 1995. <span>“<span class="nocase">Normal T lymphocytes can express two different T cell
receptor <span class="math inline">\(\beta\)</span> chains: Implications
for the mechanism of allelic exclusion</span>.”</span> <em>Journal of
Experimental Medicine</em>. <a href="https://doi.org/10.1084/jem.181.4.1587" class="external-link">https://doi.org/10.1084/jem.181.4.1587</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Andrew McDavid, Yu Gu, Erik VonKaenel, Aaron Wagner.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
