<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Repertoire with Expression with SingleCellExperiment • CellaRepertorium</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Repertoire with Expression with SingleCellExperiment">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cdr3_clustering.html">Clustering repertoire via CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amcdavid/CellaRepertorium">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Combining Repertoire with Expression with SingleCellExperiment</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/master/vignettes/repertoire_and_expression.Rmd"><code>vignettes/repertoire_and_expression.Rmd</code></a></small>
      <div class="hidden name"><code>repertoire_and_expression.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(CellaRepertorium)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; Loading required package: SummarizedExperiment</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; Loading required package: GenomicRanges</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">#&gt; The following object is masked from 'package:CellaRepertorium':</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#&gt;     rbind</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter,</span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25"><span class="co">#&gt;     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,</span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="co">#&gt;     mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span></a>
<a class="sourceLine" id="cb1-27" data-line-number="27"><span class="co">#&gt;     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,</span></a>
<a class="sourceLine" id="cb1-28" data-line-number="28"><span class="co">#&gt;     setdiff, sort, table, tapply, union, unique, unsplit, which,</span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29"><span class="co">#&gt;     which.max, which.min</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb1-31" data-line-number="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-32" data-line-number="32"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-34" data-line-number="34"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-35" data-line-number="35"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb1-36" data-line-number="36"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb1-37" data-line-number="37"><span class="co">#&gt; Loading required package: GenomeInfoDb</span></a>
<a class="sourceLine" id="cb1-38" data-line-number="38"><span class="co">#&gt; Loading required package: Biobase</span></a>
<a class="sourceLine" id="cb1-39" data-line-number="39"><span class="co">#&gt; Welcome to Bioconductor</span></a>
<a class="sourceLine" id="cb1-40" data-line-number="40"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-41" data-line-number="41"><span class="co">#&gt;     Vignettes contain introductory material; view with</span></a>
<a class="sourceLine" id="cb1-42" data-line-number="42"><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></a>
<a class="sourceLine" id="cb1-43" data-line-number="43"><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></a>
<a class="sourceLine" id="cb1-44" data-line-number="44"><span class="co">#&gt; Loading required package: DelayedArray</span></a>
<a class="sourceLine" id="cb1-45" data-line-number="45"><span class="co">#&gt; Loading required package: matrixStats</span></a>
<a class="sourceLine" id="cb1-46" data-line-number="46"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-47" data-line-number="47"><span class="co">#&gt; Attaching package: 'matrixStats'</span></a>
<a class="sourceLine" id="cb1-48" data-line-number="48"><span class="co">#&gt; The following objects are masked from 'package:Biobase':</span></a>
<a class="sourceLine" id="cb1-49" data-line-number="49"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-50" data-line-number="50"><span class="co">#&gt;     anyMissing, rowMedians</span></a>
<a class="sourceLine" id="cb1-51" data-line-number="51"><span class="co">#&gt; Loading required package: BiocParallel</span></a>
<a class="sourceLine" id="cb1-52" data-line-number="52"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-53" data-line-number="53"><span class="co">#&gt; Attaching package: 'DelayedArray'</span></a>
<a class="sourceLine" id="cb1-54" data-line-number="54"><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></a>
<a class="sourceLine" id="cb1-55" data-line-number="55"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-56" data-line-number="56"><span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></a>
<a class="sourceLine" id="cb1-57" data-line-number="57"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-58" data-line-number="58"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-59" data-line-number="59"><span class="co">#&gt;     aperm, apply, rowsum</span></a>
<a class="sourceLine" id="cb1-60" data-line-number="60"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb1-61" data-line-number="61"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-62" data-line-number="62"><span class="co">#&gt; Attaching package: 'dplyr'</span></a>
<a class="sourceLine" id="cb1-63" data-line-number="63"><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></a>
<a class="sourceLine" id="cb1-64" data-line-number="64"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-65" data-line-number="65"><span class="co">#&gt;     count</span></a>
<a class="sourceLine" id="cb1-66" data-line-number="66"><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></a>
<a class="sourceLine" id="cb1-67" data-line-number="67"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-68" data-line-number="68"><span class="co">#&gt;     combine</span></a>
<a class="sourceLine" id="cb1-69" data-line-number="69"><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></a>
<a class="sourceLine" id="cb1-70" data-line-number="70"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-71" data-line-number="71"><span class="co">#&gt;     intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb1-72" data-line-number="72"><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></a>
<a class="sourceLine" id="cb1-73" data-line-number="73"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-74" data-line-number="74"><span class="co">#&gt;     intersect</span></a>
<a class="sourceLine" id="cb1-75" data-line-number="75"><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></a>
<a class="sourceLine" id="cb1-76" data-line-number="76"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-77" data-line-number="77"><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></a>
<a class="sourceLine" id="cb1-78" data-line-number="78"><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb1-79" data-line-number="79"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-80" data-line-number="80"><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb1-81" data-line-number="81"><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></a>
<a class="sourceLine" id="cb1-82" data-line-number="82"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-83" data-line-number="83"><span class="co">#&gt;     combine, intersect, setdiff, union</span></a>
<a class="sourceLine" id="cb1-84" data-line-number="84"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-85" data-line-number="85"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-86" data-line-number="86"><span class="co">#&gt;     filter, lag</span></a>
<a class="sourceLine" id="cb1-87" data-line-number="87"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-88" data-line-number="88"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-89" data-line-number="89"><span class="co">#&gt;     intersect, setdiff, setequal, union</span></a>
<a class="sourceLine" id="cb1-90" data-line-number="90"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb1-91" data-line-number="91"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(readr)</a>
<a class="sourceLine" id="cb1-92" data-line-number="92"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyr)</a>
<a class="sourceLine" id="cb1-93" data-line-number="93"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-94" data-line-number="94"><span class="co">#&gt; Attaching package: 'tidyr'</span></a>
<a class="sourceLine" id="cb1-95" data-line-number="95"><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb1-96" data-line-number="96"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-97" data-line-number="97"><span class="co">#&gt;     expand</span></a>
<a class="sourceLine" id="cb1-98" data-line-number="98"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(stringr)</a>
<a class="sourceLine" id="cb1-99" data-line-number="99"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(purrr)</a>
<a class="sourceLine" id="cb1-100" data-line-number="100"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-101" data-line-number="101"><span class="co">#&gt; Attaching package: 'purrr'</span></a>
<a class="sourceLine" id="cb1-102" data-line-number="102"><span class="co">#&gt; The following object is masked from 'package:DelayedArray':</span></a>
<a class="sourceLine" id="cb1-103" data-line-number="103"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-104" data-line-number="104"><span class="co">#&gt;     simplify</span></a>
<a class="sourceLine" id="cb1-105" data-line-number="105"><span class="co">#&gt; The following object is masked from 'package:GenomicRanges':</span></a>
<a class="sourceLine" id="cb1-106" data-line-number="106"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-107" data-line-number="107"><span class="co">#&gt;     reduce</span></a>
<a class="sourceLine" id="cb1-108" data-line-number="108"><span class="co">#&gt; The following object is masked from 'package:IRanges':</span></a>
<a class="sourceLine" id="cb1-109" data-line-number="109"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-110" data-line-number="110"><span class="co">#&gt;     reduce</span></a></code></pre></div>
<div id="load-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#load-expression" class="anchor"></a>“Load” expression</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1345</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(ccdb_ex)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">barcodes =<span class="st"> </span>ccdb_ex<span class="op">$</span>cell_tbl[ccdb_ex<span class="op">$</span>cell_pk]</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># Take a subsample of almost all of the barcdes</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">barcodes =<span class="st"> </span>barcodes[<span class="kw"><a href="https://dplyr.tidyverse.org/reference/sample.html">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(barcodes), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(barcodes) <span class="op">-</span><span class="st"> </span><span class="dv">5</span>),]</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">samples =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(ccdb_ex<span class="op">$</span>cell_tbl[<span class="kw"><a href="https://dplyr.tidyverse.org/reference/setops.html">setdiff</a></span>(ccdb_ex<span class="op">$</span>cell_pk, <span class="st">'barcode'</span>)])</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co"># For each sample, generate  0-100 "extra" barcodes for which only 5' expression is recovered</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10">extra =<span class="st"> </span>samples <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">extrabc =</span> {</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">  extra_bc =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">floor</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">100</span>))</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">tibble</a></span>(<span class="dt">barcode =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">'barcode'</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_len</a></span>(extra_bc))))</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">}) </a>
<a class="sourceLine" id="cb2-14" data-line-number="14">extra =<span class="st"> </span>extra <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/unnest.html">unnest</a></span>()</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">all_bc =<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(extra, barcodes)</a></code></pre></div>
<p>Simulate some “cells” and “genes” that nearly form a superset of the cells for which repertoire are available. This is generally true if no barcode filters have been applied to the expression data. In practice a few cells may have repertoire but not expression (or fail QC for expression). We will work with the intersection of these cells.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">genes =<span class="st"> </span><span class="dv">200</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">cells =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(all_bc)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">array_size =<span class="st"> </span>genes<span class="op">*</span>cells</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">expression =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/NegBinomial">rnbinom</a></span>(array_size, <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">mu =</span> <span class="dv">3</span>), <span class="dt">nrow =</span> genes, <span class="dt">ncol =</span> cells)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">sce =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment">SingleCellExperiment</a></span>(<span class="dt">assay =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">counts =</span> expression), <span class="dt">colData =</span> all_bc)</a></code></pre></div>
</div>
<div id="remake-the-contigcelldb-with-empty-cells" class="section level1">
<h1 class="hasAnchor">
<a href="#remake-the-contigcelldb-with-empty-cells" class="anchor"></a>Remake the ContigCellDB with empty cells</h1>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">ccdb2 =<span class="st"> </span><span class="kw"><a href="../reference/ContigCellDB-fun.html">ContigCellDB</a></span>(ccdb_ex<span class="op">$</span>contig_tbl, <span class="dt">contig_pk =</span> ccdb_ex<span class="op">$</span>contig_pk, <span class="dt">cell_tbl =</span> <span class="kw">colData</span>(sce), <span class="dt">cell_pk =</span> ccdb_ex<span class="op">$</span>cell_pk, <span class="dt">equalize =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">ccdb2 =<span class="st"> </span><span class="kw"><a href="../reference/cdhit_ccdb.html">cdhit_ccdb</a></span>(ccdb2, <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">cluster_name =</span> <span class="st">'aa80'</span>, <span class="dt">identity =</span> <span class="fl">.8</span>, <span class="dt">min_length =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">ccdb2 =<span class="st"> </span><span class="kw"><a href="../reference/fine_clustering.html">fine_clustering</a></span>(ccdb2, <span class="dt">sequence_key =</span> <span class="st">'cdr3'</span>, <span class="dt">type =</span> <span class="st">'AA'</span>, <span class="dt">keep_clustering_details =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; Calculating intradistances on 997 clusters.</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; Summarizing</span></a></code></pre></div>
<p>Key is to construct with <code>equalize = FALSE</code>, which will allow some cells to lack contigs.</p>
</div>
<div id="chain-pairings" class="section level1">
<h1 class="hasAnchor">
<a href="#chain-pairings" class="anchor"></a>Chain pairings</h1>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">colData</span>(sce)<span class="op">$</span>alpha =<span class="st">  </span><span class="kw"><a href="../reference/canonicalize_cell.html">canonicalize_cell</a></span>(ccdb2, chain <span class="op">==</span><span class="st"> 'TRA'</span>, <span class="dt">contig_fields =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'chain'</span>, <span class="st">'v_gene'</span>,<span class="st">'d_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'aa80'</span>))</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">colData</span>(sce)<span class="op">$</span>beta =<span class="st">  </span><span class="kw"><a href="../reference/canonicalize_cell.html">canonicalize_cell</a></span>(ccdb2, chain <span class="op">==</span><span class="st"> 'TRB'</span>, <span class="dt">contig_fields =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'chain'</span>, <span class="st">'v_gene'</span>,<span class="st">'d_gene'</span>, <span class="st">'j_gene'</span>, <span class="st">'aa80'</span>))</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">colData</span>(sce)<span class="op">$</span>pairing =<span class="st"> </span><span class="kw"><a href="../reference/enumerate_pairing.html">enumerate_pairing</a></span>(ccdb2, <span class="dt">chain_recode_fun =</span> <span class="st">'guess'</span>)</a></code></pre></div>
</div>
<div id="visualization-of-tcr-features-with-scater" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization-of-tcr-features-with-scater" class="anchor"></a>Visualization of TCR features with Scater</h1>
<p>We can leverage Scater’s ability to use “nested” data frames to visualize TCR features.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(scater)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; Attaching package: 'scater'</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt;     arrange, filter, mutate, rename</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt;     rename</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; The following object is masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt;     filter</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13">sce =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalizeSCE</a></span>(sce)</a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt; Warning in normalizeSCE(sce): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15">sce =<span class="st"> </span><span class="kw">runPCA</span>(sce)</a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(sce, <span class="dt">use_dimred =</span> <span class="st">'PCA'</span>, <span class="dt">colour_by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'pairing'</span>, <span class="st">'pairing'</span>))</a></code></pre></div>
<p><img src="repertoire_and_expression_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Here we calculate the first two principal components (which don’t show much because these are simulated data without any special structure), and then visualize if the TCR was paired or not.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">only_paired =<span class="st"> </span>sce[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(sce<span class="op">$</span>pairing<span class="op">$</span>pairing <span class="op">==</span><span class="st"> 'paired'</span>)]</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(only_paired, <span class="dt">use_dimred =</span> <span class="st">'PCA'</span>, <span class="dt">colour_by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'alpha'</span>, <span class="st">'j_gene'</span>))</a></code></pre></div>
<p><img src="repertoire_and_expression_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plotReducedDim">plotReducedDim</a></span>(only_paired, <span class="dt">use_dimred =</span> <span class="st">'PCA'</span>, <span class="dt">colour_by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'beta'</span>, <span class="st">'j_gene'</span>))</a></code></pre></div>
<p><img src="repertoire_and_expression_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<p>Since the <code>ContigCellDB</code> is nested within the <code>SingleCellExperiment</code> it automatically gets subsetted appropriately when the parent object is subsetted. Enough <code>data.frame</code>-like semantics have been implemented so that fields from the <code>cell_tbl</code> can be visualized.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-expression">“Load” expression</a></li>
      <li><a href="#remake-the-contigcelldb-with-empty-cells">Remake the ContigCellDB with empty cells</a></li>
      <li><a href="#chain-pairings">Chain pairings</a></li>
      <li><a href="#visualization-of-tcr-features-with-scater">Visualization of TCR features with Scater</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Yu Gu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
