<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>R interface to CDHIT/CDHITest — cdhit • CellaRepertorium</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="R interface to CDHIT/CDHITest — cdhit"><meta property="og:description" content="CDHIT is a greedy algorithm to cluster amino acid or DNA sequences based on a
minimum identity.
By default, in this package it is configured perform ungapped, global
alignments with no clipping at start or end.
The identity is the number of identical characters in alignment
divided by the full length of the shorter sequence.
Set s &amp;lt; 1 to change the minimum coverage of the shorter sequence, which
will allow clipping at start or end.
Changing G = 0 changes the meaning of the identity to be the number of
identical characters in the alignment divided by the length of the alignment.
In this case, you must also set the alignment coverage controls aL, AL, aS, AS."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellaRepertorium</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cdr3_clustering.html">Clustering and differential usage of repertoire CDR3 sequences</a>
    </li>
    <li>
      <a href="../articles/cr-overview.html">An Introduction to CellaRepertorium</a>
    </li>
    <li>
      <a href="../articles/mouse_tcell_qc.html">Quality control and Exploration of UMI-based repertoire data</a>
    </li>
    <li>
      <a href="../articles/repertoire_and_expression.html">Combining Repertoire with Expression with SingleCellExperiment</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/amcdavid/CellaRepertorium/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R interface to CDHIT/CDHITest</h1>
    <small class="dont-index">Source: <a href="https://github.com/amcdavid/CellaRepertorium/blob/HEAD/R/cdhit-methods.R" class="external-link"><code>R/cdhit-methods.R</code></a></small>
    <div class="hidden name"><code>cdhit.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>CDHIT is a greedy algorithm to cluster amino acid or DNA sequences based on a
minimum identity.
By default, in this package it is configured perform ungapped, global
alignments with no clipping at start or end.
The <code>identity</code> is the number of identical characters in alignment
divided by the full length of the shorter sequence.
Set <code>s</code> &lt; 1 to change the minimum coverage of the shorter sequence, which
will allow clipping at start or end.
Changing <code>G</code> = 0 changes the meaning of the <code>identity</code> to be the number of
identical characters in the alignment divided by the length of the alignment.
In this case, you must also set the alignment coverage controls <code>aL</code>, <code>AL</code>, <code>aS</code>, <code>AS</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">cdhit</span><span class="op">(</span>
  <span class="va">seqs</span>,
  identity <span class="op">=</span> <span class="cn">NULL</span>,
  kmerSize <span class="op">=</span> <span class="cn">NULL</span>,
  min_length <span class="op">=</span> <span class="fl">6</span>,
  s <span class="op">=</span> <span class="fl">1</span>,
  G <span class="op">=</span> <span class="fl">1</span>,
  only_index <span class="op">=</span> <span class="cn">FALSE</span>,
  showProgress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seqs</dt>
<dd><p><code>AAseq</code> or <code>DNAseq</code></p></dd>
<dt>identity</dt>
<dd><p>minimum proportion identity</p></dd>
<dt>kmerSize</dt>
<dd><p>word size.  If NULL, it will be chosen automatically based on
the identity.
You may need to lower it below 5 for AAseq with identity less than .7.</p></dd>
<dt>min_length</dt>
<dd><p>Minimum length for sequences to be clustered.  An error if
something smaller is passed.</p></dd>
<dt>s</dt>
<dd><p>fraction of shorter sequence covered by alignment.</p></dd>
<dt>G</dt>
<dd><p>1 for global alignment, 0 for local. If doubt, pick global.</p></dd>
<dt>only_index</dt>
<dd><p>if TRUE only return the integer cluster indices, otherwise
return a tibble.</p></dd>
<dt>showProgress</dt>
<dd><p>show a status bar</p></dd>
<dt>...</dt>
<dd><p>other arguments that can be passed to cdhit, see
https://github.com/weizhongli/cdhit/wiki/3.-User's-Guide#CDHIT for details.
These will override any default values.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>vector of <code>integer</code> of length <code>seqs</code> providing the cluster
ID for each sequence, or a <code>tibble</code>.  See details.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>CDHit is by Fu, Niu, Zhu, Wu and Li (2012).  The R interface is originally by
Thomas Lin Pedersen and was transcribed here because it is not exported from
the package FindMyFriends, which is orphaned.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">fasta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'demo.fasta'</span>, package<span class="op">=</span><span class="st">'CellaRepertorium'</span><span class="op">)</span></span>
<span class="r-in"><span class="va">aaseq</span> <span class="op">=</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readAAStringSet</a></span><span class="op">(</span><span class="va">fasta_path</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># 100% identity, global alignment</span></span>
<span class="r-in"><span class="fu">cdhit</span><span class="op">(</span><span class="va">aaseq</span>, identity <span class="op">=</span> <span class="fl">1</span>, only_index <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 100 101 162 102   6 245 103  49 163 164</span>
<span class="r-in"><span class="co"># 100% identity, local alignment with no padding of endpoints</span></span>
<span class="r-in"><span class="fu">cdhit</span><span class="op">(</span><span class="va">aaseq</span>,identity <span class="op">=</span> <span class="fl">1</span>, G <span class="op">=</span> <span class="fl">0</span>, aL <span class="op">=</span> <span class="fl">1</span>, aS <span class="op">=</span> <span class="fl">1</span>,  only_index <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 100 101 162 102   6 245 103  49 163 164</span>
<span class="r-in"><span class="co"># 100% identity, local alignment with .9 padding of endpoints</span></span>
<span class="r-in"><span class="fu">cdhit</span><span class="op">(</span><span class="va">aaseq</span>,identity <span class="op">=</span> <span class="fl">1</span>, G <span class="op">=</span> <span class="fl">0</span>, aL <span class="op">=</span> <span class="fl">.9</span>, aS <span class="op">=</span> <span class="fl">.9</span>,  only_index <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 100 101 162 102   6 245 103  49 163 164</span>
<span class="r-in"><span class="co"># a tibble</span></span>
<span class="r-in"><span class="va">tbl</span> <span class="op">=</span> <span class="fu">cdhit</span><span class="op">(</span><span class="va">aaseq</span>, identity <span class="op">=</span> <span class="fl">1</span>, G <span class="op">=</span> <span class="fl">0</span>, aL <span class="op">=</span> <span class="fl">.9</span>, aS <span class="op">=</span> <span class="fl">.9</span>, only_index <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Andrew McDavid, Yu Gu, Erik VonKaenel, Aaron Wagner.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

